{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da6bc368-f4fa-46ae-9e40-a03a90e7e125",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff3dd304-0ffa-4566-8216-fc21b838afea",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import squarify\n",
    "import pandas as pd\n",
    "import rpy2.robjects as ro\n",
    "import matplotlib\n",
    "\n",
    "from collections import defaultdict\n",
    "from copy import deepcopy\n",
    "from scipy import stats\n",
    "from colour import Color\n",
    "from matplotlib import pyplot as plt\n",
    "from pathlib import Path\n",
    "from upsetplot import UpSet, from_contents\n",
    "from sklearn.decomposition import PCA\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib_venn import venn3, venn3_circles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43a9cf7a-330e-47c3-8a13-036ad19c78bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "src_path: str = \"/home/uziel/Development/biopipes/src\"\n",
    "sys.path.insert(0, src_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a68f68a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from r_wrappers.utils import rpy2_df_to_pd_df, pd_df_to_rpy2_df\n",
    "from r_wrappers.deseq2 import vst_transform\n",
    "from r_wrappers.limma import *\n",
    "from components.functional_analysis.orgdb import OrgDB\n",
    "\n",
    "org_db = OrgDB(\"Homo sapiens\")\n",
    "root = Path(\"/home/uziel/storage\")\n",
    "pcta_wcdt_root = root.joinpath(\"PCTA_WCDT_FOLH1\")\n",
    "tcga_root = root.joinpath(\"TCGA_PRAD_MethArray_FOLH1\")\n",
    "wcdt_root = root.joinpath(\"WCDT_MCRPC_FOLH1\")\n",
    "palette_str = \"flare\"\n",
    "\n",
    "sns.set_theme(style=\"whitegrid\", palette=palette_str)\n",
    "plt.rcParams[\"font.family\"] = \"serif\"\n",
    "plt.rcParams[\"font.serif\"] = \"Ubuntu Mono\"\n",
    "plt.rcParams[\"font.monospace\"] = \"Ubuntu Mono\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c944600e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pcta_wcdt_figs_path = pcta_wcdt_root.joinpath(\"manuscript_figures\")\n",
    "pcta_wcdt_figs_path.mkdir(parents=True, exist_ok=True)\n",
    "pcta_wcdt_add_files_path = pcta_wcdt_root.joinpath(\"additional_files\")\n",
    "pcta_wcdt_add_files_path.mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa191ac3",
   "metadata": {},
   "outputs": [],
   "source": [
    "pcta_wcdt_annot_df = pd.read_csv(\n",
    "    pcta_wcdt_root.joinpath(\"data\").joinpath(\"samples_annotation_FOLH1.csv\"),\n",
    "    index_col=0,\n",
    ")\n",
    "tcga_annot_df = pd.read_csv(\n",
    "    tcga_root.joinpath(\"data\").joinpath(\"samples_annotation_FOLH1.csv\"), index_col=0\n",
    ")\n",
    "wcdt_annot_df = pd.read_csv(\n",
    "    wcdt_root.joinpath(\"data\").joinpath(\"samples_annotation_FOLH1.csv\"), index_col=0\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a3babfbc",
   "metadata": {},
   "source": [
    "## Figure 1: Workflow\n",
    "\n",
    "A workflow schematic showing our method from start to finish. Kind of a graphical abstract, it must be visually attractive and contain few technical details. **This should give the reader an overview of the whole methodology.**\n",
    "\n",
    "\n",
    "![Methodology Workflow](../../../docs/1_workflow.svg)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9bdec83a",
   "metadata": {},
   "source": [
    "## Figure 2: Data\n",
    "\n",
    "Focus on PCTA+WCDT (PCTA+), but mention that a subset of samples has DNA methylation or WGBS data available. **This should show a good summary of the data used in the study.**\n",
    "\n",
    "1. (**A**) PCA of all samples (before batch correction)\n",
    "2. (**B**) Treemap with dataset composition of PCTA+\n",
    "3. (**C**) Treemap of tissue origin in metastatic samples\n",
    "4. (**D**) PCA of all samples (after batch correction)\n",
    "5. (**E**) PSMA levels of each group in a raincloud plot (prim)\n",
    "6. (**F**) PSMA levels of each group in a raincloud plot (met)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with pd.ExcelWriter(pcta_wcdt_add_files_path / \"additional_file_1.xlsx\") as writer:\n",
    "    pcta_wcdt_annot_df.to_excel(writer, sheet_name=\"1A. PCTA+ Annotations\")\n",
    "    tcga_annot_df.to_excel(writer, sheet_name=\"1B. TCGA-PRAD Annotations\")\n",
    "    wcdt_annot_df.to_excel(writer, sheet_name=\"1C. WCDT-MCRPC Annotations\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "235d991f",
   "metadata": {},
   "source": [
    "### 2A: PCA of all samples (before batch correction)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d1ed1a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_df = pd.read_csv(\n",
    "    root.joinpath(\"PCTA_WCDT\").joinpath(\"data\").joinpath(\"raw_counts.csv\"), index_col=0\n",
    ")\n",
    "print(counts_df.shape)\n",
    "counts_df_wo_batch_effects = pd.read_csv(\n",
    "    root.joinpath(\"PCTA_WCDT\")\n",
    "    .joinpath(\"data\")\n",
    "    .joinpath(\"raw_counts_wo_batch_effects.csv\"),\n",
    "    index_col=0,\n",
    ")\n",
    "print(counts_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bb16308",
   "metadata": {},
   "outputs": [],
   "source": [
    "common_samples = pcta_wcdt_annot_df.index.intersection(counts_df.columns)\n",
    "annot_df_contrasts = pcta_wcdt_annot_df.loc[common_samples, :]\n",
    "counts_df = counts_df.loc[:, common_samples]\n",
    "counts_df_wo_batch_effects = counts_df_wo_batch_effects.loc[:, common_samples]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d465054b",
   "metadata": {},
   "outputs": [],
   "source": [
    "vst_df = rpy2_df_to_pd_df(\n",
    "    vst_transform(\n",
    "        ro.r(\"as.matrix\")(pd_df_to_rpy2_df(counts_df.loc[counts_df.mean(axis=1) > 1]))\n",
    "    )\n",
    ")\n",
    "vst_wo_batch_effects_df = rpy2_df_to_pd_df(\n",
    "    vst_transform(\n",
    "        ro.r(\"as.matrix\")(\n",
    "            pd_df_to_rpy2_df(\n",
    "                counts_df_wo_batch_effects.loc[\n",
    "                    counts_df_wo_batch_effects.mean(axis=1) > 10\n",
    "                ]\n",
    "            )\n",
    "        )\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bec4ff61",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 4), dpi=100)\n",
    "\n",
    "pca = PCA(n_components=2)\n",
    "pca_comps = pca.fit_transform(vst_df.transpose().values)\n",
    "palette = {\n",
    "    \"norm\": \"#9ACD32\",\n",
    "    \"prim\": \"#4A708B\",\n",
    "    \"met\": \"#8B3A3A\",\n",
    "}\n",
    "\n",
    "annot_df_contrasts[\"Sample Type\"] = annot_df_contrasts[\"sample_type\"]\n",
    "annot_df_contrasts[\"Library Type\"] = annot_df_contrasts[\"library_type\"].map(\n",
    "    {\"total_rna\": \"Total RNA\", \"hybrid_capture\": \"Hybrid Capture\", \"poly_a\": \"Poly(A)\"}\n",
    ")\n",
    "sns.scatterplot(\n",
    "    x=pca_comps.T[0],\n",
    "    y=pca_comps.T[1],\n",
    "    hue=annot_df_contrasts[\"Sample Type\"],\n",
    "    palette=palette,\n",
    "    style=annot_df_contrasts[\"Library Type\"],\n",
    "    ax=ax,\n",
    ")\n",
    "\n",
    "plt.legend(loc=\"upper right\")\n",
    "# plt.suptitle(\"PCA of PCTA+WCDT samples (before batch correction)\", fontsize=18)\n",
    "pca_variances = pca.explained_variance_ratio_\n",
    "ax.set_xlabel(f\"PC1 ({pca_variances[0] * 100:.2f}%)\", fontsize=14)\n",
    "ax.set_ylabel(f\"PC2 ({pca_variances[1] * 100:.2f}%)\", fontsize=14)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"2A_pca_before_combat.pdf\"), bbox_inches=\"tight\"\n",
    ")\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"2A_pca_before_combat.svg\"), bbox_inches=\"tight\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "681042c8",
   "metadata": {},
   "source": [
    "### 2B: Treemap with dataset composition of PCTA+"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6df1f5c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 4), dpi=100)\n",
    "\n",
    "data = pcta_wcdt_annot_df\n",
    "sizes = data[\"dataset\"].value_counts()\n",
    "\n",
    "labels = [f\"{label}\\n({value}/{len(data)})\" for label, value in sizes.items()]\n",
    "\n",
    "norm = matplotlib.colors.Normalize(vmin=sizes.values.min(), vmax=sizes.values.max())\n",
    "cmap = sns.color_palette(\"flare_r\", as_cmap=True)\n",
    "colors = [cmap(norm(value)) for value in sizes.values]\n",
    "\n",
    "squarify.plot(\n",
    "    sizes=sizes.values,\n",
    "    label=labels,\n",
    "    alpha=0.8,\n",
    "    color=colors,\n",
    "    text_kwargs={\"fontsize\": 10},\n",
    ")\n",
    "# plt.title(\"PCTA+ dataset composition\")\n",
    "plt.axis(\"off\")\n",
    "plt.tight_layout()\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"2B_dataset_origin.pdf\"), bbox_inches=\"tight\")\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"2B_dataset_origin.svg\"), bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfe2213f",
   "metadata": {},
   "source": [
    "### 2C: Treemap of tissue origin in metastatic samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b92c3bbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 4), dpi=100)\n",
    "\n",
    "data = pcta_wcdt_annot_df[pcta_wcdt_annot_df[\"sample_type\"] == \"met\"]\n",
    "data.loc[:, \"tumor_tissue_site\"] = data[\"tumor_tissue_site\"].apply(\n",
    "    lambda x: x.replace(\"_\", \" \").title() if isinstance(x, str) else x\n",
    ")\n",
    "sizes = data[\"tumor_tissue_site\"].value_counts()\n",
    "sizes_clean = sizes.drop(sizes[sizes < 7].index)\n",
    "sizes_clean[\"Other\"] = sum(sizes[sizes < 7])\n",
    "\n",
    "labels = [f\"{label}\\n({value}/{len(data)})\" for label, value in sizes_clean.items()]\n",
    "\n",
    "norm = matplotlib.colors.Normalize(\n",
    "    vmin=sizes_clean.values.min(), vmax=sizes_clean.values.max()\n",
    ")\n",
    "cmap = sns.color_palette(\"flare_r\", as_cmap=True)\n",
    "colors = [cmap(norm(value)) for value in sizes_clean.values]\n",
    "\n",
    "squarify.plot(\n",
    "    sizes=sizes_clean.values,\n",
    "    label=labels,\n",
    "    alpha=0.8,\n",
    "    color=colors,\n",
    "    text_kwargs={\"fontsize\": 10},\n",
    ")\n",
    "# plt.title(\"Origin tissue site of metastatic samples in PCTA+\")\n",
    "plt.axis(\"off\")\n",
    "plt.tight_layout()\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"2C_tissue_origin.pdf\"), bbox_inches=\"tight\")\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"2C_tissue_origin.svg\"), bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e9685b8",
   "metadata": {},
   "source": [
    "### 2D: PCA of all samples (after batch correction)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b18ac35c",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1, figsize=(8, 4), dpi=100)\n",
    "\n",
    "pca = PCA(n_components=2)\n",
    "pca_comps = pca.fit_transform(vst_wo_batch_effects_df.transpose().values)\n",
    "palette = {\n",
    "    \"norm\": \"#9ACD32\",\n",
    "    \"prim\": \"#4A708B\",\n",
    "    \"met\": \"#8B3A3A\",\n",
    "}\n",
    "annot_df_contrasts[\"Sample Type\"] = annot_df_contrasts[\"sample_type\"]\n",
    "annot_df_contrasts[\"Library Type\"] = annot_df_contrasts[\"library_type\"].map(\n",
    "    {\"total_rna\": \"Total RNA\", \"hybrid_capture\": \"Hybrid Capture\", \"poly_a\": \"Poly(A)\"}\n",
    ")\n",
    "sns.scatterplot(\n",
    "    x=pca_comps.T[0],\n",
    "    y=pca_comps.T[1],\n",
    "    hue=annot_df_contrasts[\"Sample Type\"],\n",
    "    palette=palette,\n",
    "    style=annot_df_contrasts[\"Library Type\"],\n",
    "    ax=ax,\n",
    ")\n",
    "\n",
    "plt.legend(loc=\"lower right\")\n",
    "# plt.suptitle(\"PCA of PCTA+WCDT samples (after batch correction)\", fontsize=18)\n",
    "pca_variances = pca.explained_variance_ratio_\n",
    "ax.set_xlabel(f\"PC1 ({pca_variances[0] * 100:.2f}%)\", fontsize=14)\n",
    "ax.set_ylabel(f\"PC2 ({pca_variances[1] * 100:.2f}%)\", fontsize=14)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"2D_pca_after_combat.pdf\"), bbox_inches=\"tight\"\n",
    ")\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"2D_pca_after_combat.svg\"), bbox_inches=\"tight\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2E: PSMA levels of each group in a raincloud plot (prim)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_df = pcta_wcdt_annot_df[pcta_wcdt_annot_df[\"sample_type\"] == \"prim\"]\n",
    "\n",
    "data_x = []\n",
    "folh1_levels = []\n",
    "for folh1_level in (\"high\", \"mid\", \"low\"):\n",
    "    x = data_df[data_df[\"FOLH1_level\"] == folh1_level][\"FOLH1_VST\"]\n",
    "    data_x.append(x)\n",
    "    folh1_levels.append(folh1_level + f\"\\n({len(x)})\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1, figsize=(8, 4), dpi=100)\n",
    "\n",
    "# Boxplot data\n",
    "bp = ax.boxplot(data_x, patch_artist=True, vert=False, widths=0.1)\n",
    "\n",
    "# Change to the desired color and add transparency\n",
    "for patch in bp[\"boxes\"]:\n",
    "    # patch.set_facecolor(color)\n",
    "    patch.set_alpha(0.4)\n",
    "\n",
    "# Violinplot data\n",
    "vp = ax.violinplot(\n",
    "    data_x,\n",
    "    # points=500,\n",
    "    showmeans=False,\n",
    "    showextrema=False,\n",
    "    showmedians=False,\n",
    "    vert=False,\n",
    ")\n",
    "\n",
    "for idx, b in enumerate(vp[\"bodies\"]):\n",
    "    # Get the center of the plot\n",
    "    m = np.mean(b.get_paths()[0].vertices[:, 0])\n",
    "\n",
    "    # Modify it so we only see the upper half of the violin plot\n",
    "    b.get_paths()[0].vertices[:, 1] = np.clip(\n",
    "        b.get_paths()[0].vertices[:, 1], idx + 1, idx + 2\n",
    "    )\n",
    "\n",
    "# Scatterplot data\n",
    "for idx, features in enumerate(data_x):\n",
    "    # Add jitter effect so the features do not overlap on the y-axis\n",
    "    y = np.full(len(features), idx + 0.8)\n",
    "    idxs = np.arange(len(y))\n",
    "    out = y.astype(float)\n",
    "    out.flat[idxs] += np.random.uniform(low=-0.05, high=0.05, size=len(idxs))\n",
    "    y = out\n",
    "    plt.scatter(features, y, s=1)\n",
    "\n",
    "plt.yticks(np.arange(1, len(folh1_levels) + 1, 1), folh1_levels)\n",
    "plt.ylabel(\"FOLH1 group\", fontsize=14)\n",
    "plt.xlabel(\"FOLH1 RNA Expression (VST-transformed)\", fontsize=14)\n",
    "# plt.suptitle(\"Optimal stratification of primary samples in PCTA+\", fontsize=18)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"2E_prim_psma_rainclouds.pdf\"), bbox_inches=\"tight\"\n",
    ")\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"2E_prim_psma_rainclouds.svg\"), bbox_inches=\"tight\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3710c349",
   "metadata": {},
   "source": [
    "### 2F: PSMA levels of each group in a raincloud plot (met)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e10e1cbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_df = pcta_wcdt_annot_df[pcta_wcdt_annot_df[\"sample_type\"] == \"met\"]\n",
    "\n",
    "data_x = []\n",
    "folh1_levels = []\n",
    "for folh1_level in (\"high\", \"mid\", \"low\"):\n",
    "    x = data_df[data_df[\"FOLH1_level\"] == folh1_level][\"FOLH1_VST\"]\n",
    "    data_x.append(x)\n",
    "    folh1_levels.append(folh1_level + f\"\\n({len(x)})\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8eb045d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(8, 4), dpi=100)\n",
    "\n",
    "# Boxplot data\n",
    "bp = ax.boxplot(data_x, patch_artist=True, vert=False, widths=0.1)\n",
    "\n",
    "# Change to the desired color and add transparency\n",
    "for patch in bp[\"boxes\"]:\n",
    "    # patch.set_facecolor(color)\n",
    "    patch.set_alpha(0.4)\n",
    "\n",
    "# Violinplot data\n",
    "vp = ax.violinplot(\n",
    "    data_x,\n",
    "    # points=500,\n",
    "    showmeans=False,\n",
    "    showextrema=False,\n",
    "    showmedians=False,\n",
    "    vert=False,\n",
    ")\n",
    "\n",
    "for idx, b in enumerate(vp[\"bodies\"]):\n",
    "    # Get the center of the plot\n",
    "    m = np.mean(b.get_paths()[0].vertices[:, 0])\n",
    "\n",
    "    # Modify it so we only see the upper half of the violin plot\n",
    "    b.get_paths()[0].vertices[:, 1] = np.clip(\n",
    "        b.get_paths()[0].vertices[:, 1], idx + 1, idx + 2\n",
    "    )\n",
    "\n",
    "# Scatterplot data\n",
    "for idx, features in enumerate(data_x):\n",
    "    # Add jitter effect so the features do not overlap on the y-axis\n",
    "    y = np.full(len(features), idx + 0.8)\n",
    "    idxs = np.arange(len(y))\n",
    "    out = y.astype(float)\n",
    "    out.flat[idxs] += np.random.uniform(low=-0.05, high=0.05, size=len(idxs))\n",
    "    y = out\n",
    "    plt.scatter(features, y, s=1)\n",
    "\n",
    "plt.yticks(np.arange(1, len(folh1_levels) + 1, 1), folh1_levels)\n",
    "plt.ylabel(\"FOLH1 group\", fontsize=14)\n",
    "plt.xlabel(\"FOLH1 RNA Expression (VST-transformed)\", fontsize=14)\n",
    "# plt.suptitle(\"Optimal stratification of metastatic samples in PCTA+\", fontsize=18)\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"2F_met_psma_rainclouds.pdf\"), bbox_inches=\"tight\"\n",
    ")\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"2F_met_psma_rainclouds.svg\"), bbox_inches=\"tight\"\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "111e86f8",
   "metadata": {},
   "source": [
    "## Figure 3: Differential analyses\n",
    "\n",
    "This figure should summarize the gene sets obtained from D-GSVA and GSEA (DE & DM).\n",
    "\n",
    "1. (**A**) Barplot of D-GSVA gene sets for category H (prim over/under)\n",
    "2. (**B**) Barplot of D-GSVA gene sets for category H (met over/under)\n",
    "3. (**C**) Starburst plot (prim)\n",
    "4. (**D**) Barplot of GSEA (DE) gene sets for category H (prim over/under)\n",
    "5. (**E**) Barplot of GSEA (DM) gene sets for category H (prim over/under)\n",
    "6. (**F**) Starburst plot (met)\n",
    "7. (**G**) Barplot of GSEA (DE) gene sets for category H (met over/under)\n",
    "8. (**H**) Barplot of GSEA (DM) gene sets for category H (met over/under)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "writer_2 = pd.ExcelWriter(pcta_wcdt_add_files_path / \"additional_file_2.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3A: Barplot of D-GSVA gene sets for category H (prim over/under)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dgsva_h_df = pd.read_csv(\n",
    "    pcta_wcdt_root.joinpath(\"diff_gsva\")\n",
    "    .joinpath(\"H\")\n",
    "    .joinpath(\n",
    "        \"sample_type_met+prim_FOLH1_level_prim_high+prim_low_\"\n",
    "        \"_prim_high_vs_prim_low_top_table_padj_0_05_all_0_0.csv\"\n",
    "    ),\n",
    "    index_col=0,\n",
    ")\n",
    "dgsva_h_df.index = [\n",
    "    idx.replace(\"HALLMARK_\", \"\").replace(\"_\", \" \") for idx in dgsva_h_df.index\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dgsva_h_df.to_excel(writer_2, sheet_name=\"2A. D-GSVA H (PRIM)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = dgsva_h_df[\"log2FoldChange\"].sort_values(ascending=False).round(2).to_frame()\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(14, 10), dpi=100)\n",
    "\n",
    "ax = sns.barplot(data=data, x=\"log2FoldChange\", y=data.index, palette=palette_str)\n",
    "for container in ax.containers:\n",
    "    ax.bar_label(container, padding=5)\n",
    "\n",
    "ax.set_ylabel(\"Gene Sets\", fontsize=20)\n",
    "plt.yticks(fontsize=16)\n",
    "ax.set_xlabel(\"$Log_2$ Fold Change\", fontsize=20)\n",
    "ax.set_xlim([ax.get_xlim()[0] * 1.2, ax.get_xlim()[1] * 1.2])\n",
    "# plt.title(\n",
    "#     \"D-GSVA Analysis on MSigDB H Collection (Prim. High vs Prim. low)\",\n",
    "#     fontsize=20,\n",
    "#     y=1.02,\n",
    "# )\n",
    "plt.tight_layout()\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"3A_dgsva_h_prim.pdf\"), bbox_inches=\"tight\")\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"3A_dgsva_h_prim.svg\"), bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3B: Barplot of D-GSVA gene sets for category H (met over/under)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dgsva_h_df = pd.read_csv(\n",
    "    pcta_wcdt_root.joinpath(\"diff_gsva\")\n",
    "    .joinpath(\"H\")\n",
    "    .joinpath(\n",
    "        \"sample_type_met+prim_FOLH1_level_met_high+met_low_\"\n",
    "        \"_met_high_vs_met_low_top_table_padj_0_05_all_0_0.csv\"\n",
    "    ),\n",
    "    index_col=0,\n",
    ")\n",
    "dgsva_h_df.index = [\n",
    "    idx.replace(\"HALLMARK_\", \"\").replace(\"_\", \" \") for idx in dgsva_h_df.index\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dgsva_h_df.to_excel(writer_2, sheet_name=\"2B. D-GSVA H (MET)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = dgsva_h_df[\"log2FoldChange\"].sort_values(ascending=False).round(2).to_frame()\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(14, 10), dpi=100)\n",
    "\n",
    "ax = sns.barplot(data=data, x=\"log2FoldChange\", y=data.index, palette=palette_str)\n",
    "for container in ax.containers:\n",
    "    ax.bar_label(container, padding=5)\n",
    "\n",
    "ax.set_ylabel(\"Gene Sets\", fontsize=20)\n",
    "plt.yticks(fontsize=16)\n",
    "ax.set_xlabel(\"$Log_2$ Fold Change\", fontsize=20)\n",
    "ax.set_xlim([ax.get_xlim()[0] * 1.2, ax.get_xlim()[1] * 1.2])\n",
    "# plt.title(\n",
    "#     \"D-GSVA Analysis on MSigDB H Collection (Met. High vs Met. low)\",\n",
    "#     fontsize=20,\n",
    "#     y=1.02,\n",
    "# )\n",
    "plt.tight_layout()\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"3B_dgsva_h_met.pdf\"), bbox_inches=\"tight\")\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"3B_dgsva_h_met.svg\"), bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24d54e5f",
   "metadata": {},
   "source": [
    "### 3C: Starburst plot (prim)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a5c6a77",
   "metadata": {},
   "outputs": [],
   "source": [
    "label_colors = {\n",
    "    \"Not relevant\": \"gray\",\n",
    "    \"Up regulated\": \"#94090D\",\n",
    "    \"Up regulated & Hyper methylated\": \"#450003\",\n",
    "    \"Up regulated & Hypo methylated\": \"#FF1D23\",\n",
    "    \"Down regulated\": \"#A7E7F6\",\n",
    "    \"Down regulated & Hyper methylated\": \"#318EAD\",\n",
    "    \"Down regulated & Hypo methylated\": \"#225A76\",\n",
    "    \"Hyper methylated\": \"#F06E19\",\n",
    "    \"Hypo methylated\": \"#3FCEF6\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e53f9fc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\n",
    "    (\n",
    "        pcta_wcdt_root.joinpath(\"deseq2\").joinpath(\n",
    "            \"sample_type_met+prim_FOLH1_level_prim_high+prim_low__\"\n",
    "            \"prim_high_vs_prim_low_deseq_results_unique.csv\"\n",
    "        )\n",
    "    ),\n",
    "    index_col=0,\n",
    ").set_index(\"SYMBOL\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cfb6132",
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_expr_df = pd.read_csv(\n",
    "    (\n",
    "        pcta_wcdt_root.joinpath(\"deseq2\").joinpath(\n",
    "            \"sample_type_met+prim_FOLH1_level_prim_high+prim_low__\"\n",
    "            \"prim_high_vs_prim_low_deseq_results_unique.csv\"\n",
    "        )\n",
    "    ),\n",
    "    index_col=0,\n",
    ").set_index(\"SYMBOL\")\n",
    "diff_expr_df = diff_expr_df[diff_expr_df[\"padj\"] < 0.05]\n",
    "\n",
    "diff_meth_df = (\n",
    "    pd.read_csv(\n",
    "        (\n",
    "            tcga_root.joinpath(\"minfi\").joinpath(\n",
    "                \"sample_type_prim_FOLH1_level_prim_high+prim_low__\"\n",
    "                \"diff_meth_probes_noob_quantile_top_table_prim_high_vs_prim_low_\"\n",
    "                \"wrt_mean_diff_0_0_hg38_genes_promoters.csv\"\n",
    "            )\n",
    "        ),\n",
    "        index_col=0,\n",
    "    )\n",
    "    .replace(\"NA_character_\", np.nan)\n",
    "    .dropna(subset=[\"annot.symbol\"])\n",
    ")\n",
    "diff_meth_df = (\n",
    "    diff_meth_df[diff_meth_df[\"adj.P.Val\"] < 0.05]\n",
    "    .groupby(\"annot.symbol\")\n",
    "    .mean(numeric_only=True)\n",
    "    .rename_axis(\"SYMBOL\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cceacdc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "common_genes = diff_expr_df.index.intersection(diff_meth_df.index)\n",
    "print(len(common_genes))\n",
    "starburst_df = pd.DataFrame()\n",
    "starburst_df[\"expr_lfc\"] = diff_expr_df.loc[common_genes, \"log2FoldChange\"]\n",
    "starburst_df[\"meth_lfc\"] = diff_meth_df.loc[common_genes, \"logFC\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec5266b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "starburst_df.loc[\n",
    "    (abs(starburst_df[\"expr_lfc\"]) < 1) & (abs(starburst_df[\"meth_lfc\"]) < 1), \"label\"\n",
    "] = \"Not relevant\"\n",
    "\n",
    "starburst_df.loc[\n",
    "    (starburst_df[\"expr_lfc\"] > 1) & (abs(starburst_df[\"meth_lfc\"]) < 1), \"label\"\n",
    "] = \"Up regulated\"\n",
    "starburst_df.loc[\n",
    "    (starburst_df[\"expr_lfc\"] > 1) & (starburst_df[\"meth_lfc\"] > 1), \"label\"\n",
    "] = \"Up regulated & Hyper methylated\"\n",
    "starburst_df.loc[\n",
    "    (starburst_df[\"expr_lfc\"] > 1) & (starburst_df[\"meth_lfc\"] < -1), \"label\"\n",
    "] = \"Up regulated & Hypo methylated\"\n",
    "\n",
    "starburst_df.loc[\n",
    "    (starburst_df[\"expr_lfc\"] < -1) & (abs(starburst_df[\"meth_lfc\"]) < 1), \"label\"\n",
    "] = \"Down regulated\"\n",
    "starburst_df.loc[\n",
    "    (starburst_df[\"expr_lfc\"] < -1) & (starburst_df[\"meth_lfc\"] > 1), \"label\"\n",
    "] = \"Down regulated & Hyper methylated\"\n",
    "starburst_df.loc[\n",
    "    (starburst_df[\"expr_lfc\"] < -1) & (starburst_df[\"meth_lfc\"] < -1), \"label\"\n",
    "] = \"Down regulated & Hypo methylated\"\n",
    "\n",
    "starburst_df.loc[\n",
    "    (abs(starburst_df[\"expr_lfc\"]) < 1) & (starburst_df[\"meth_lfc\"] > 1), \"label\"\n",
    "] = \"Hyper methylated\"\n",
    "starburst_df.loc[\n",
    "    (abs(starburst_df[\"expr_lfc\"]) < 1) & (starburst_df[\"meth_lfc\"] < -1), \"label\"\n",
    "] = \"Hypo methylated\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "starburst_df.to_excel(writer_2, sheet_name=\"2C. Starburst (PRIM)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f5c38ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "prim_label_colors = {\n",
    "    k: v for k, v in label_colors.items() if k in starburst_df[\"label\"].unique()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bf0f7a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(14, 10), dpi=100)\n",
    "\n",
    "ax = sns.scatterplot(\n",
    "    data=starburst_df,\n",
    "    x=\"expr_lfc\",\n",
    "    y=\"meth_lfc\",\n",
    "    hue=\"label\",\n",
    "    palette=prim_label_colors,\n",
    "    hue_order=prim_label_colors.keys(),\n",
    "    ax=ax,\n",
    "    s=15,\n",
    ")\n",
    "highlight_labels = (\n",
    "    \"Up regulated & Hyper methylated\",\n",
    "    \"Up regulated & Hypo methylated\",\n",
    "    \"Down regulated & Hyper methylated\",\n",
    "    \"Down regulated & Hypo methylated\",\n",
    ")\n",
    "\n",
    "labels_counter = dict()\n",
    "labels_text_y = defaultdict(lambda: 99)\n",
    "for gene, meta in starburst_df.sort_values(\n",
    "    [\"expr_lfc\", \"meth_lfc\"], key=abs, ascending=False\n",
    ").iterrows():\n",
    "    label = meta[\"label\"]\n",
    "    if label in highlight_labels:\n",
    "        if labels_counter.get(label, 0) == 5:\n",
    "            continue\n",
    "        labels_counter[label] = labels_counter.get(label, 0) + 1\n",
    "\n",
    "        text_y_new = meta[\"meth_lfc\"] + 0.1\n",
    "        text_y_new = (\n",
    "            text_y_new\n",
    "            if abs(labels_text_y[label] - text_y_new) > 0.1\n",
    "            else (labels_text_y[label] + 0.1)\n",
    "        )\n",
    "        labels_text_y[label] = text_y_new\n",
    "\n",
    "        plt.annotate(\n",
    "            gene,\n",
    "            xy=(meta[\"expr_lfc\"], meta[\"meth_lfc\"]),\n",
    "            xytext=(meta[\"expr_lfc\"] + 0.25, text_y_new),\n",
    "            arrowprops=dict(facecolor=\"black\", width=2, headwidth=5, headlength=5),\n",
    "            fontsize=14,\n",
    "            weight=\"semibold\",\n",
    "        )\n",
    "\n",
    "ymin, ymax = (\n",
    "    starburst_df[\"meth_lfc\"].min() - 0.5,\n",
    "    starburst_df[\"meth_lfc\"].max() + 0.5,\n",
    ")\n",
    "plt.ylim((ymin, ymax))\n",
    "\n",
    "xmin, xmax = (\n",
    "    starburst_df[\"expr_lfc\"].min() - 0.5,\n",
    "    starburst_df[\"expr_lfc\"].max() + 0.5,\n",
    ")\n",
    "plt.xlim((xmin, xmax))\n",
    "\n",
    "# vertical lines\n",
    "plt.vlines(\n",
    "    1,\n",
    "    ymin=ymin,\n",
    "    ymax=ymax,\n",
    "    colors=\"grey\",\n",
    "    linestyles=\"dashed\",\n",
    ")\n",
    "plt.vlines(\n",
    "    -1,\n",
    "    ymin=ymin,\n",
    "    ymax=ymax,\n",
    "    colors=\"grey\",\n",
    "    linestyles=\"dashed\",\n",
    ")\n",
    "\n",
    "# horizontal lines\n",
    "plt.hlines(\n",
    "    1,\n",
    "    xmin=xmin,\n",
    "    xmax=xmax,\n",
    "    colors=\"grey\",\n",
    "    linestyles=\"dashed\",\n",
    ")\n",
    "plt.hlines(\n",
    "    -1,\n",
    "    xmin=xmin,\n",
    "    xmax=xmax,\n",
    "    colors=\"grey\",\n",
    "    linestyles=\"dashed\",\n",
    ")\n",
    "\n",
    "# axis labels\n",
    "plt.xlabel(\"Gene Expression ($Log_2$ Fold Change)\", fontsize=20)\n",
    "plt.ylabel(\"DNA Methylation ($Log_2$ Fold Change)\", fontsize=20)\n",
    "# plt.legend(title=r\"$\\bf Multi-omics\\ relationship$\", loc=(1.02, 0.3))\n",
    "plt.legend(title=r\"$\\bf Multi-omics\\ relationship$\", loc=\"lower left\")\n",
    "# plt.suptitle(\n",
    "#     \"Multi-omics gene signatures (prim. high vs prim. low)\",\n",
    "#     fontsize=20,\n",
    "# )\n",
    "plt.tight_layout()\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"3C_prim_starburst.pdf\"), bbox_inches=\"tight\")\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"3C_prim_starburst.svg\"), bbox_inches=\"tight\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "cd48365a",
   "metadata": {},
   "source": [
    "### 3D: Barplot of GSEA (DE) gene sets for category H (prim over/under)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7107eeae",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_h_df = pd.read_csv(\n",
    "    pcta_wcdt_root.joinpath(\"functional\")\n",
    "    .joinpath(\"MSIGDB\")\n",
    "    .joinpath(\n",
    "        \"sample_type_met+prim_FOLH1_level_prim_high+prim_low_\"\n",
    "        \"_prim_high_vs_prim_low_gsea_H.csv\"\n",
    "    ),\n",
    "    index_col=0,\n",
    ")\n",
    "gsea_h_df.index = [\n",
    "    idx.replace(\"HALLMARK_\", \"\").replace(\"_\", \" \") for idx in gsea_h_df.index\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_h_df.to_excel(writer_2, sheet_name=\"2D. GSEA DE H (PRIM)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e253d58",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = gsea_h_df[\"enrichmentScore\"].sort_values(ascending=False).round(2).to_frame()\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(14, 10), dpi=100)\n",
    "\n",
    "ax = sns.barplot(data=data, x=\"enrichmentScore\", y=data.index, palette=palette_str)\n",
    "for container in ax.containers:\n",
    "    ax.bar_label(container, padding=5)\n",
    "\n",
    "ax.set_ylabel(\"Gene Sets\", fontsize=20)\n",
    "plt.yticks(fontsize=16)\n",
    "ax.set_xlabel(\"Enrichment Score\", fontsize=20)\n",
    "ax.set_xlim([ax.get_xlim()[0] * 1.2, ax.get_xlim()[1] * 1.2])\n",
    "# plt.title(\n",
    "#     \"GSEA (DE) Analysis on MSigDB H Collection (Prim. High vs Prim. low)\",\n",
    "#     fontsize=20,\n",
    "#     y=1.02,\n",
    "# )\n",
    "plt.tight_layout()\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"3D_gsea_de_h_prim.pdf\"), bbox_inches=\"tight\")\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"3D_gsea_de_h_prim.svg\"), bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3E: Barplot of GSEA (DM) gene sets for category H (prim over/under)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_h_df = pd.read_csv(\n",
    "    tcga_root.joinpath(\"functional\")\n",
    "    .joinpath(\"MSIGDB\")\n",
    "    .joinpath(\n",
    "        \"sample_type_prim_FOLH1_level_prim_high+prim_low_\"\n",
    "        \"_diff_meth_probes_noob_quantile_top_table_prim_high_vs_prim_low_\"\n",
    "        \"wrt_mean_diff_0_0_hg38_genes_promoters_gsea_H.csv\"\n",
    "    ),\n",
    "    index_col=0,\n",
    ")\n",
    "gsea_h_df.index = [\n",
    "    idx.replace(\"HALLMARK_\", \"\").replace(\"_\", \" \") for idx in gsea_h_df.index\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_h_df.to_excel(writer_2, sheet_name=\"2E. GSEA DM H (PRIM)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = gsea_h_df[\"enrichmentScore\"].sort_values(ascending=False).round(2).to_frame()\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(14, 10), dpi=100)\n",
    "\n",
    "ax = sns.barplot(data=data, x=\"enrichmentScore\", y=data.index, palette=palette_str)\n",
    "for container in ax.containers:\n",
    "    ax.bar_label(container, padding=5)\n",
    "\n",
    "ax.set_ylabel(\"Gene Sets\", fontsize=20)\n",
    "plt.yticks(fontsize=16)\n",
    "ax.set_xlabel(\"Enrichment Score\", fontsize=20)\n",
    "ax.set_xlim([ax.get_xlim()[0] * 1.2, ax.get_xlim()[1] * 1.2])\n",
    "# plt.title(\n",
    "#     \"GSEA (DM) Analysis on MSigDB H Collection (Prim. High vs Prim. low)\",\n",
    "#     fontsize=20,\n",
    "#     y=1.02,\n",
    "# )\n",
    "plt.tight_layout()\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"3E_gsea_dm_h_prim.pdf\"), bbox_inches=\"tight\")\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"3E_gsea_dm_h_prim.svg\"), bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9234a9e",
   "metadata": {},
   "source": [
    "### 3F: Starburst plot (met)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b88aa890",
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_expr_df = pd.read_csv(\n",
    "    (\n",
    "        pcta_wcdt_root.joinpath(\"deseq2\").joinpath(\n",
    "            \"sample_type_met+prim_FOLH1_level_met_high+met_low__\"\n",
    "            \"met_high_vs_met_low_deseq_results_unique.csv\"\n",
    "        )\n",
    "    ),\n",
    "    index_col=0,\n",
    ").set_index(\"SYMBOL\")\n",
    "diff_expr_df = diff_expr_df[diff_expr_df[\"padj\"] < 0.05]\n",
    "\n",
    "diff_meth_df = (\n",
    "    pd.read_csv(\n",
    "        (\n",
    "            wcdt_root.joinpath(\"methylkit\").joinpath(\n",
    "                \"sample_type_met_FOLH1_level_met_high+met_low__\"\n",
    "                \"diff_meth_met_high_vs_met_low_ann_genes_hg38_genes_promoters.csv\"\n",
    "            )\n",
    "        ),\n",
    "        index_col=0,\n",
    "    )\n",
    "    .replace(\"NA_character_\", np.nan)\n",
    "    .dropna(subset=[\"annot.symbol\"])\n",
    ")\n",
    "diff_meth_df = (\n",
    "    diff_meth_df[diff_meth_df[\"qvalue\"] < 0.05]\n",
    "    .groupby(\"annot.symbol\")\n",
    "    .mean(numeric_only=True)\n",
    "    .rename_axis(\"SYMBOL\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eba8ec2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "common_genes = diff_expr_df.index.intersection(diff_meth_df.index)\n",
    "print(len(common_genes))\n",
    "starburst_df = pd.DataFrame()\n",
    "starburst_df[\"expr_lfc\"] = diff_expr_df.loc[common_genes, \"log2FoldChange\"]\n",
    "starburst_df[\"meth_diff\"] = diff_meth_df.loc[common_genes, \"meth.diff\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c80e162",
   "metadata": {},
   "outputs": [],
   "source": [
    "meth_diff_th = 15"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f56af81",
   "metadata": {},
   "outputs": [],
   "source": [
    "starburst_df.loc[\n",
    "    (abs(starburst_df[\"expr_lfc\"]) < 1)\n",
    "    & (abs(starburst_df[\"meth_diff\"]) < meth_diff_th),\n",
    "    \"label\",\n",
    "] = \"Not relevant\"\n",
    "\n",
    "starburst_df.loc[\n",
    "    (starburst_df[\"expr_lfc\"] > 1) & (abs(starburst_df[\"meth_diff\"]) < meth_diff_th),\n",
    "    \"label\",\n",
    "] = \"Up regulated\"\n",
    "starburst_df.loc[\n",
    "    (starburst_df[\"expr_lfc\"] > 1) & (starburst_df[\"meth_diff\"] > meth_diff_th), \"label\"\n",
    "] = \"Up regulated & Hyper methylated\"\n",
    "starburst_df.loc[\n",
    "    (starburst_df[\"expr_lfc\"] > 1) & (starburst_df[\"meth_diff\"] < -meth_diff_th),\n",
    "    \"label\",\n",
    "] = \"Up regulated & Hypo methylated\"\n",
    "\n",
    "starburst_df.loc[\n",
    "    (starburst_df[\"expr_lfc\"] < -1) & (abs(starburst_df[\"meth_diff\"]) < meth_diff_th),\n",
    "    \"label\",\n",
    "] = \"Down regulated\"\n",
    "starburst_df.loc[\n",
    "    (starburst_df[\"expr_lfc\"] < -1) & (starburst_df[\"meth_diff\"] > meth_diff_th),\n",
    "    \"label\",\n",
    "] = \"Down regulated & Hyper methylated\"\n",
    "starburst_df.loc[\n",
    "    (starburst_df[\"expr_lfc\"] < -1) & (starburst_df[\"meth_diff\"] < -meth_diff_th),\n",
    "    \"label\",\n",
    "] = \"Down regulated & Hypo methylated\"\n",
    "\n",
    "starburst_df.loc[\n",
    "    (abs(starburst_df[\"expr_lfc\"]) < 1) & (starburst_df[\"meth_diff\"] > meth_diff_th),\n",
    "    \"label\",\n",
    "] = \"Hyper methylated\"\n",
    "starburst_df.loc[\n",
    "    (abs(starburst_df[\"expr_lfc\"]) < 1) & (starburst_df[\"meth_diff\"] < -meth_diff_th),\n",
    "    \"label\",\n",
    "] = \"Hypo methylated\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f8315b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "starburst_df.to_excel(writer_2, sheet_name=\"2D. Starburst (MET)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d15c341",
   "metadata": {},
   "outputs": [],
   "source": [
    "met_label_colors = {\n",
    "    k: v for k, v in label_colors.items() if k in starburst_df[\"label\"].unique()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4109c3ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(14, 10), dpi=100)\n",
    "\n",
    "ax = sns.scatterplot(\n",
    "    data=starburst_df,\n",
    "    x=\"expr_lfc\",\n",
    "    y=\"meth_diff\",\n",
    "    hue=\"label\",\n",
    "    palette=met_label_colors,\n",
    "    hue_order=met_label_colors.keys(),\n",
    "    ax=ax,\n",
    "    s=15,\n",
    ")\n",
    "highlight_labels = (\n",
    "    \"Up regulated & Hyper methylated\",\n",
    "    \"Up regulated & Hypo methylated\",\n",
    "    \"Down regulated & Hyper methylated\",\n",
    "    \"Down regulated & Hypo methylated\",\n",
    ")\n",
    "\n",
    "labels_counter = dict()\n",
    "labels_text_y = defaultdict(lambda: 99)\n",
    "for gene, meta in starburst_df.sort_values(\n",
    "    [\"expr_lfc\", \"meth_diff\"], key=abs, ascending=False\n",
    ").iterrows():\n",
    "    label = meta[\"label\"]\n",
    "    if label in highlight_labels:\n",
    "        if labels_counter.get(label, 0) == 5:\n",
    "            continue\n",
    "        labels_counter[label] = labels_counter.get(label, 0) + 1\n",
    "\n",
    "        text_y_new = meta[\"meth_diff\"] - 1.5\n",
    "        text_y_new = (\n",
    "            text_y_new\n",
    "            if abs(labels_text_y[label] - text_y_new) > 1.5\n",
    "            else (labels_text_y[label] - 1.5)\n",
    "        )\n",
    "        labels_text_y[label] = text_y_new\n",
    "\n",
    "        plt.annotate(\n",
    "            gene,\n",
    "            xy=(meta[\"expr_lfc\"], meta[\"meth_diff\"]),\n",
    "            xytext=(meta[\"expr_lfc\"] + 0.5, text_y_new),\n",
    "            arrowprops=dict(facecolor=\"black\", width=2, headwidth=5, headlength=5),\n",
    "            fontsize=14,\n",
    "            weight=\"semibold\",\n",
    "        )\n",
    "\n",
    "ymin, ymax = (\n",
    "    starburst_df[\"meth_diff\"].min() - 0.5,\n",
    "    starburst_df[\"meth_diff\"].max() + 0.5,\n",
    ")\n",
    "plt.ylim((ymin, ymax))\n",
    "\n",
    "xmin, xmax = (\n",
    "    starburst_df[\"expr_lfc\"].min() - 0.5,\n",
    "    starburst_df[\"expr_lfc\"].max() + 0.5,\n",
    ")\n",
    "plt.xlim((xmin, xmax))\n",
    "\n",
    "# vertical lines\n",
    "plt.vlines(\n",
    "    1,\n",
    "    ymin=ymin,\n",
    "    ymax=ymax,\n",
    "    colors=\"grey\",\n",
    "    linestyles=\"dashed\",\n",
    ")\n",
    "plt.vlines(\n",
    "    -1,\n",
    "    ymin=ymin,\n",
    "    ymax=ymax,\n",
    "    colors=\"grey\",\n",
    "    linestyles=\"dashed\",\n",
    ")\n",
    "\n",
    "# horizontal lines\n",
    "plt.hlines(\n",
    "    meth_diff_th,\n",
    "    xmin=xmin,\n",
    "    xmax=xmax,\n",
    "    colors=\"grey\",\n",
    "    linestyles=\"dashed\",\n",
    ")\n",
    "plt.hlines(\n",
    "    -meth_diff_th,\n",
    "    xmin=xmin,\n",
    "    xmax=xmax,\n",
    "    colors=\"grey\",\n",
    "    linestyles=\"dashed\",\n",
    ")\n",
    "\n",
    "# axis labels\n",
    "plt.xlabel(\"Gene Expression ($Log_2$ Fold Change)\", fontsize=20)\n",
    "plt.ylabel(\"DNA Methylation (Methylation difference)\", fontsize=20)\n",
    "# plt.legend(title=r\"$\\bf Multi-omics\\ relationship$\", loc=(1.02, 0.3))\n",
    "plt.legend(title=r\"$\\bf Multi-omics\\ relationship$\", loc=\"lower left\")\n",
    "# plt.suptitle(\n",
    "#     \"Multi-omics gene signatures (MET. high vs MET. low)\",\n",
    "#     fontsize=20,\n",
    "# )\n",
    "plt.tight_layout()\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"3F_met_starburst.pdf\"), bbox_inches=\"tight\")\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"3F_met_starburst.svg\"), bbox_inches=\"tight\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "d09cf9d6",
   "metadata": {},
   "source": [
    "### 3G: Barplot of GSEA (DE) gene sets for category H (met over/under)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a3a11e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_h_df = pd.read_csv(\n",
    "    pcta_wcdt_root.joinpath(\"functional\")\n",
    "    .joinpath(\"MSIGDB\")\n",
    "    .joinpath(\n",
    "        \"sample_type_met+prim_FOLH1_level_met_high+met_low_\"\n",
    "        \"_met_high_vs_met_low_gsea_H.csv\"\n",
    "    ),\n",
    "    index_col=0,\n",
    ")\n",
    "gsea_h_df.index = [\n",
    "    idx.replace(\"HALLMARK_\", \"\").replace(\"_\", \" \") for idx in gsea_h_df.index\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gsea_h_df.to_excel(writer_2, sheet_name=\"2G. GSEA DE H (MET)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "982780e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = gsea_h_df[\"enrichmentScore\"].sort_values(ascending=False).round(2).to_frame()\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(14, 10), dpi=100)\n",
    "\n",
    "ax = sns.barplot(data=data, x=\"enrichmentScore\", y=data.index, palette=palette_str)\n",
    "for container in ax.containers:\n",
    "    ax.bar_label(container, padding=5)\n",
    "\n",
    "ax.set_ylabel(\"Gene Sets\", fontsize=20)\n",
    "plt.yticks(fontsize=16)\n",
    "ax.set_xlabel(\"Enrichment Score\", fontsize=20)\n",
    "ax.set_xlim([ax.get_xlim()[0] * 1.2, ax.get_xlim()[1] * 1.2])\n",
    "# plt.title(\n",
    "#     \"GSEA (DE) Analysis on MSigDB H Collection (Met. High vs Met. low)\",\n",
    "#     fontsize=20,\n",
    "#     y=1.02,\n",
    "# )\n",
    "plt.tight_layout()\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"3G_gsea_de_h_met.pdf\"), bbox_inches=\"tight\")\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"3G_gsea_de_h_met.svg\"), bbox_inches=\"tight\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9c866fb7",
   "metadata": {},
   "source": [
    "### 3H: Barplot of GSEA (DM) gene sets for category H (met over/under)\n",
    "\n",
    "**No results available!**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "writer_2.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98cb1d46",
   "metadata": {},
   "source": [
    "## Figure 4: Identifying PSMA protein community\n",
    "\n",
    "This figure summarizes the sets of top genes found and the PPI network analysis.\n",
    "\n",
    "1. (**A**) Jointplot of z-scores vs. log2 fold changes of DEGs (prim over)\n",
    "2. (**B**) Jointplot of z-scores vs. log2 fold changes of DMGs (prim over)\n",
    "3. (**C**) Jointplot of z-scores vs. log2 fold changes of DEGs (prim under)\n",
    "4. (**D**) Jointplot of z-scores vs. log2 fold changes of DMGs (prim under)\n",
    "5. (**E**) Jointplot of z-scores vs. log2 fold changes of DEGs (met over)\n",
    "6. (**F**) Jointplot of z-scores vs. log2 fold changes of DMGs (met over)\n",
    "7. (**G**) Jointplot of z-scores vs. log2 fold changes of DEGs (met under)\n",
    "8. (**H**) Jointplot of z-scores vs. log2 fold changes of DMGs (met under)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "writer_3 = pd.ExcelWriter(pcta_wcdt_add_files_path / \"additional_file_3.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1015dc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ymin, ymax = (-1, 12)\n",
    "xmin, xmax = (-8, 8)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "55d01278",
   "metadata": {},
   "source": [
    "### 4A. Jointplot of z-scores vs. log2 fold changes of DEGs (prim over)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e3dd22d",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_genes_df = pd.read_csv(\n",
    "    pcta_wcdt_root.joinpath(\"degss_genes_stats\")\n",
    "    .joinpath(\n",
    "        \"prim_high_vs_prim_low_degss_padj_0_05_up_0_0_\"\n",
    "        \"gsea_H_C1_C2_C3_C4_C5_C6_C7_C8_union\"\n",
    "    )\n",
    "    .joinpath(\"gene_occurrence_stats_bootstrap_z_score_summary_ann.csv\"),\n",
    "    index_col=0,\n",
    ")\n",
    "print(top_genes_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_genes_df.to_excel(writer_3, sheet_name=\"3A. Top genes stats (PRIM over)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dd787ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "x, y, hue = [\"diff_expr_log2FoldChange\", \"mean_z_score\", \"Statistical significance\"]\n",
    "expr_scatter_df = top_genes_df[[x, y]].dropna()\n",
    "\n",
    "expr_scatter_df[hue] = \"Significant\"\n",
    "expr_scatter_df.loc[expr_scatter_df[\"mean_z_score\"] < 1.96, hue] = \"Not Significant\"\n",
    "expr_scatter_df.loc[abs(expr_scatter_df[x]) < 1, hue] = \"Not Significant\"\n",
    "\n",
    "palette = {\"Significant\": \"#8B3A3A\", \"Not Significant\": \"#BFBFBF\"}\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6, 5), dpi=100)\n",
    "\n",
    "grid = sns.scatterplot(data=expr_scatter_df, x=x, y=y, hue=hue, palette=palette, ax=ax)\n",
    "\n",
    "top_up = (\n",
    "    expr_scatter_df.loc[\n",
    "        (expr_scatter_df[x] > 1) & (expr_scatter_df[hue] == \"Significant\")\n",
    "    ]\n",
    "    .sort_values(y, ascending=False)\n",
    "    .head(5)\n",
    ")\n",
    "top_down = (\n",
    "    expr_scatter_df.loc[\n",
    "        (expr_scatter_df[x] < -1) & (expr_scatter_df[hue] == \"Significant\")\n",
    "    ]\n",
    "    .sort_values(y, ascending=False)\n",
    "    .head(5)\n",
    ")\n",
    "\n",
    "text_y = 99\n",
    "for gene, gene_meta in top_up.iterrows():\n",
    "    text_y_new = gene_meta[y] + 0.75\n",
    "    text_y_new = (\n",
    "        (text_y := text_y_new)\n",
    "        if text_y - text_y_new > 0.75\n",
    "        else (text_y := text_y - 0.75)\n",
    "    )\n",
    "    plt.annotate(\n",
    "        gene,\n",
    "        xy=(gene_meta[x], gene_meta[y]),\n",
    "        xytext=(gene_meta[x] + 2, text_y_new),\n",
    "        arrowprops=dict(facecolor=\"black\", width=2, headwidth=5, headlength=5),\n",
    "        fontsize=12,\n",
    "    )\n",
    "\n",
    "text_y = 99\n",
    "for gene, gene_meta in top_down.iterrows():\n",
    "    text_y_new = gene_meta[y] + 0.75\n",
    "    text_y_new = (\n",
    "        (text_y := text_y_new)\n",
    "        if text_y - text_y_new > 0.75\n",
    "        else (text_y := text_y - 0.75)\n",
    "    )\n",
    "    plt.annotate(\n",
    "        gene,\n",
    "        xy=(gene_meta[x], gene_meta[y]),\n",
    "        xytext=(gene_meta[x] - 3, text_y_new),\n",
    "        arrowprops=dict(facecolor=\"black\", width=2, headwidth=5, headlength=5),\n",
    "        fontsize=12,\n",
    "    )\n",
    "\n",
    "plt.xlim(xmin, xmax)\n",
    "plt.ylim(ymin, ymax)\n",
    "plt.legend(loc=\"upper left\")\n",
    "plt.vlines(1, ymin=ymin, ymax=ymax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "plt.vlines(-1, ymin=ymin, ymax=ymax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "plt.hlines(1.96, xmin=xmin, xmax=xmax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "# plt.suptitle(\"Gene Occurrence of DEGs (Prim over)\", y=1.03)\n",
    "plt.ylabel(\"Gene occurrence (Z-score normalised)\", fontsize=14)\n",
    "plt.xlabel(r\"$Log_{2}$ Fold Change\", fontsize=14)\n",
    "plt.legend(loc=\"upper left\")\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"4A_jointplot_diff_expr_prim_over.pdf\"),\n",
    "    bbox_inches=\"tight\",\n",
    ")\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"4A_jointplot_diff_expr_prim_over.svg\"),\n",
    "    bbox_inches=\"tight\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4B. Jointplot of z-scores vs. log2 fold changes of DMGs (prim over)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bea30599",
   "metadata": {},
   "outputs": [],
   "source": [
    "x, y, hue = [\"diff_meth_promoters_logFC\", \"mean_z_score\", \"Statistical significance\"]\n",
    "meth_scatter_df = top_genes_df[[x, y]].dropna()\n",
    "\n",
    "meth_scatter_df[hue] = \"Significant\"\n",
    "meth_scatter_df.loc[meth_scatter_df[\"mean_z_score\"] < 1.96, hue] = \"Not Significant\"\n",
    "meth_scatter_df.loc[abs(meth_scatter_df[x]) < 1, hue] = \"Not Significant\"\n",
    "\n",
    "palette = {\"Significant\": \"#8B3A3A\", \"Not Significant\": \"#BFBFBF\"}\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6, 5), dpi=100)\n",
    "\n",
    "grid = sns.scatterplot(data=meth_scatter_df, x=x, y=y, hue=hue, palette=palette, ax=ax)\n",
    "\n",
    "top_up = (\n",
    "    meth_scatter_df.loc[\n",
    "        (meth_scatter_df[x] > 1) & (meth_scatter_df[hue] == \"Significant\")\n",
    "    ]\n",
    "    .sort_values(y, ascending=False)\n",
    "    .head(5)\n",
    ")\n",
    "top_down = (\n",
    "    meth_scatter_df.loc[\n",
    "        (meth_scatter_df[x] < -1) & (meth_scatter_df[hue] == \"Significant\")\n",
    "    ]\n",
    "    .sort_values(y, ascending=False)\n",
    "    .head(5)\n",
    ")\n",
    "\n",
    "text_y = 99\n",
    "for gene, gene_meta in top_up.iterrows():\n",
    "    text_y_new = gene_meta[y] + 0.75\n",
    "    text_y_new = (\n",
    "        (text_y := text_y_new)\n",
    "        if text_y - text_y_new > 0.75\n",
    "        else (text_y := text_y - 0.75)\n",
    "    )\n",
    "    plt.annotate(\n",
    "        gene,\n",
    "        xy=(gene_meta[x], gene_meta[y]),\n",
    "        xytext=(gene_meta[x] + 2, text_y_new),\n",
    "        arrowprops=dict(facecolor=\"black\", width=2, headwidth=5, headlength=5),\n",
    "        fontsize=12,\n",
    "    )\n",
    "\n",
    "text_y = 99\n",
    "for gene, gene_meta in top_down.iterrows():\n",
    "    text_y_new = gene_meta[y] + 0.75\n",
    "    text_y_new = (\n",
    "        (text_y := text_y_new)\n",
    "        if text_y - text_y_new > 0.75\n",
    "        else (text_y := text_y - 0.75)\n",
    "    )\n",
    "    plt.annotate(\n",
    "        gene,\n",
    "        xy=(gene_meta[x], gene_meta[y]),\n",
    "        xytext=(gene_meta[x] - 3, text_y_new),\n",
    "        arrowprops=dict(facecolor=\"black\", width=2, headwidth=5, headlength=5),\n",
    "        fontsize=12,\n",
    "    )\n",
    "\n",
    "\n",
    "plt.xlim(xmin, xmax)\n",
    "plt.ylim(ymin, ymax)\n",
    "plt.legend(loc=\"upper left\")\n",
    "plt.vlines(1, ymin=ymin, ymax=ymax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "plt.vlines(-1, ymin=ymin, ymax=ymax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "plt.hlines(1.96, xmin=xmin, xmax=xmax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "# plt.suptitle(\"Gene Occurrence of DMGs (Prim over)\", y=1.03)\n",
    "plt.ylabel(\"Gene occurrence (Z-score normalised)\", fontsize=14)\n",
    "plt.xlabel(r\"$Log_{2}$ Fold Change\", fontsize=14)\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"4B_jointplot_diff_meth_prim_over.pdf\"),\n",
    "    bbox_inches=\"tight\",\n",
    ")\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"4B_jointplot_diff_meth_prim_over.svg\"),\n",
    "    bbox_inches=\"tight\",\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "59a5975b",
   "metadata": {},
   "source": [
    "### 4C. Jointplot of z-scores vs. log2 fold changes of DEGs (prim under)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c242624e",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_genes_df = pd.read_csv(\n",
    "    pcta_wcdt_root.joinpath(\"degss_genes_stats\")\n",
    "    .joinpath(\n",
    "        \"prim_high_vs_prim_low_degss_padj_0_05_down_0_0_\"\n",
    "        \"gsea_H_C1_C2_C3_C4_C5_C6_C7_C8_union\"\n",
    "    )\n",
    "    .joinpath(\"gene_occurrence_stats_bootstrap_z_score_summary_ann.csv\"),\n",
    "    index_col=0,\n",
    ")\n",
    "print(top_genes_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_genes_df.to_excel(writer_3, sheet_name=\"3B. Top genes stats (PRIM under)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25c6e4be",
   "metadata": {},
   "outputs": [],
   "source": [
    "x, y, hue = [\"diff_expr_log2FoldChange\", \"mean_z_score\", \"Statistical significance\"]\n",
    "expr_scatter_df = top_genes_df[[x, y]].dropna()\n",
    "\n",
    "expr_scatter_df[hue] = \"Significant\"\n",
    "expr_scatter_df.loc[expr_scatter_df[\"mean_z_score\"] < 1.96, hue] = \"Not Significant\"\n",
    "expr_scatter_df.loc[abs(expr_scatter_df[x]) < 1, hue] = \"Not Significant\"\n",
    "\n",
    "palette = {\"Significant\": \"#8B3A3A\", \"Not Significant\": \"#BFBFBF\"}\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6, 5), dpi=100)\n",
    "\n",
    "grid = sns.scatterplot(data=expr_scatter_df, x=x, y=y, hue=hue, palette=palette, ax=ax)\n",
    "\n",
    "top_up = (\n",
    "    expr_scatter_df.loc[\n",
    "        (expr_scatter_df[x] > 1) & (expr_scatter_df[hue] == \"Significant\")\n",
    "    ]\n",
    "    .sort_values(y, ascending=False)\n",
    "    .head(5)\n",
    ")\n",
    "top_down = (\n",
    "    expr_scatter_df.loc[\n",
    "        (expr_scatter_df[x] < -1) & (expr_scatter_df[hue] == \"Significant\")\n",
    "    ]\n",
    "    .sort_values(y, ascending=False)\n",
    "    .head(5)\n",
    ")\n",
    "\n",
    "text_y = 99\n",
    "for gene, gene_meta in top_up.iterrows():\n",
    "    text_y_new = gene_meta[y] + 0.75\n",
    "    text_y_new = (\n",
    "        (text_y := text_y_new)\n",
    "        if text_y - text_y_new > 0.75\n",
    "        else (text_y := text_y - 0.75)\n",
    "    )\n",
    "    plt.annotate(\n",
    "        gene,\n",
    "        xy=(gene_meta[x], gene_meta[y]),\n",
    "        xytext=(gene_meta[x] + 2, text_y_new),\n",
    "        arrowprops=dict(facecolor=\"black\", width=2, headwidth=5, headlength=5),\n",
    "        fontsize=12,\n",
    "    )\n",
    "\n",
    "text_y = 99\n",
    "for gene, gene_meta in top_down.iterrows():\n",
    "    text_y_new = gene_meta[y] + 0.75\n",
    "    text_y_new = (\n",
    "        (text_y := text_y_new)\n",
    "        if text_y - text_y_new > 0.75\n",
    "        else (text_y := text_y - 0.75)\n",
    "    )\n",
    "    plt.annotate(\n",
    "        gene,\n",
    "        xy=(gene_meta[x], gene_meta[y]),\n",
    "        xytext=(gene_meta[x] - 3, text_y_new),\n",
    "        arrowprops=dict(facecolor=\"black\", width=2, headwidth=5, headlength=5),\n",
    "        fontsize=12,\n",
    "    )\n",
    "\n",
    "\n",
    "plt.xlim(xmin, xmax)\n",
    "plt.ylim(ymin, ymax)\n",
    "plt.legend(loc=\"upper left\")\n",
    "plt.vlines(1, ymin=ymin, ymax=ymax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "plt.vlines(-1, ymin=ymin, ymax=ymax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "plt.hlines(1.96, xmin=xmin, xmax=xmax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "# plt.suptitle(\"Gene Occurrence of DEGs (Prim under)\", y=1.03)\n",
    "plt.ylabel(\"Gene occurrence (Z-score normalised)\", fontsize=14)\n",
    "plt.xlabel(r\"$Log_{2}$ Fold Change\", fontsize=14)\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"4C_jointplot_diff_expr_prim_under.pdf\"),\n",
    "    bbox_inches=\"tight\",\n",
    ")\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"4C_jointplot_diff_expr_prim_under.svg\"),\n",
    "    bbox_inches=\"tight\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4D. Jointplot of z-scores vs. log2 fold changes of DMGs (prim under)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a9b35a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "x, y, hue = [\"diff_meth_promoters_logFC\", \"mean_z_score\", \"Statistical significance\"]\n",
    "meth_scatter_df = top_genes_df[[x, y]].dropna()\n",
    "\n",
    "meth_scatter_df[hue] = \"Significant\"\n",
    "meth_scatter_df.loc[meth_scatter_df[\"mean_z_score\"] < 1.96, hue] = \"Not Significant\"\n",
    "meth_scatter_df.loc[abs(meth_scatter_df[x]) < 1, hue] = \"Not Significant\"\n",
    "\n",
    "palette = {\"Significant\": \"#8B3A3A\", \"Not Significant\": \"#BFBFBF\"}\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6, 5), dpi=100)\n",
    "\n",
    "grid = sns.scatterplot(data=meth_scatter_df, x=x, y=y, hue=hue, palette=palette, ax=ax)\n",
    "\n",
    "top_up = (\n",
    "    meth_scatter_df.loc[\n",
    "        (meth_scatter_df[x] > 1) & (meth_scatter_df[hue] == \"Significant\")\n",
    "    ]\n",
    "    .sort_values(y, ascending=False)\n",
    "    .head(5)\n",
    ")\n",
    "top_down = (\n",
    "    meth_scatter_df.loc[\n",
    "        (meth_scatter_df[x] < -1) & (meth_scatter_df[hue] == \"Significant\")\n",
    "    ]\n",
    "    .sort_values(y, ascending=False)\n",
    "    .head(5)\n",
    ")\n",
    "\n",
    "text_y = 99\n",
    "for gene, gene_meta in top_up.iterrows():\n",
    "    text_y_new = gene_meta[y] + 0.75\n",
    "    text_y_new = (\n",
    "        (text_y := text_y_new)\n",
    "        if text_y - text_y_new > 0.75\n",
    "        else (text_y := text_y - 0.75)\n",
    "    )\n",
    "    plt.annotate(\n",
    "        gene,\n",
    "        xy=(gene_meta[x], gene_meta[y]),\n",
    "        xytext=(gene_meta[x] + 2, text_y_new),\n",
    "        arrowprops=dict(facecolor=\"black\", width=2, headwidth=5, headlength=5),\n",
    "        fontsize=12,\n",
    "    )\n",
    "\n",
    "text_y = 99\n",
    "for gene, gene_meta in top_down.iterrows():\n",
    "    text_y_new = gene_meta[y] + 0.75\n",
    "    text_y_new = (\n",
    "        (text_y := text_y_new)\n",
    "        if text_y - text_y_new > 0.75\n",
    "        else (text_y := text_y - 0.75)\n",
    "    )\n",
    "    plt.annotate(\n",
    "        gene,\n",
    "        xy=(gene_meta[x], gene_meta[y]),\n",
    "        xytext=(gene_meta[x] - 3, text_y_new),\n",
    "        arrowprops=dict(facecolor=\"black\", width=2, headwidth=5, headlength=5),\n",
    "        fontsize=12,\n",
    "    )\n",
    "\n",
    "\n",
    "plt.xlim(xmin, xmax)\n",
    "plt.ylim(ymin, ymax)\n",
    "plt.legend(loc=\"upper left\")\n",
    "plt.vlines(1, ymin=ymin, ymax=ymax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "plt.vlines(-1, ymin=ymin, ymax=ymax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "plt.hlines(1.96, xmin=xmin, xmax=xmax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "# plt.suptitle(\"Gene Occurrence of DMGs (Prim under)\", y=1.03)\n",
    "plt.ylabel(\"Gene occurrence (Z-score normalised)\", fontsize=14)\n",
    "plt.xlabel(r\"$Log_{2}$ Fold Change\", fontsize=14)\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"4D_jointplot_diff_meth_prim_under.pdf\"),\n",
    "    bbox_inches=\"tight\",\n",
    ")\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"4D_jointplot_diff_meth_prim_under.svg\"),\n",
    "    bbox_inches=\"tight\",\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a01d06e1",
   "metadata": {},
   "source": [
    "### 4E. Jointplot of z-scores vs. log2 fold changes of DEGs (met over)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f130fe8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_genes_df = pd.read_csv(\n",
    "    pcta_wcdt_root.joinpath(\"degss_genes_stats\")\n",
    "    .joinpath(\n",
    "        \"met_high_vs_met_low_degss_padj_0_05_up_0_0_\"\n",
    "        \"gsea_H_C1_C2_C3_C4_C5_C6_C7_C8_union\"\n",
    "    )\n",
    "    .joinpath(\"gene_occurrence_stats_bootstrap_z_score_summary_ann.csv\"),\n",
    "    index_col=0,\n",
    ")\n",
    "print(top_genes_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_genes_df.to_excel(writer_3, sheet_name=\"3C. Top genes stats (MET over)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce05f8cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "x, y, hue = [\"diff_expr_log2FoldChange\", \"mean_z_score\", \"Statistical significance\"]\n",
    "expr_scatter_df = top_genes_df[[x, y]].dropna()\n",
    "\n",
    "expr_scatter_df[hue] = \"Significant\"\n",
    "expr_scatter_df.loc[expr_scatter_df[\"mean_z_score\"] < 1.96, hue] = \"Not Significant\"\n",
    "expr_scatter_df.loc[abs(expr_scatter_df[x]) < 1, hue] = \"Not Significant\"\n",
    "\n",
    "palette = {\"Significant\": \"#8B3A3A\", \"Not Significant\": \"#BFBFBF\"}\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6, 5), dpi=100)\n",
    "\n",
    "grid = sns.scatterplot(data=expr_scatter_df, x=x, y=y, hue=hue, palette=palette, ax=ax)\n",
    "\n",
    "top_up = (\n",
    "    expr_scatter_df.loc[\n",
    "        (expr_scatter_df[x] > 1) & (expr_scatter_df[hue] == \"Significant\")\n",
    "    ]\n",
    "    .sort_values(y, ascending=False)\n",
    "    .head(5)\n",
    ")\n",
    "top_down = (\n",
    "    expr_scatter_df.loc[\n",
    "        (expr_scatter_df[x] < -1) & (expr_scatter_df[hue] == \"Significant\")\n",
    "    ]\n",
    "    .sort_values(y, ascending=False)\n",
    "    .head(5)\n",
    ")\n",
    "\n",
    "text_y = 99\n",
    "for gene, gene_meta in top_up.iterrows():\n",
    "    text_y_new = gene_meta[y] + 0.75\n",
    "    text_y_new = (\n",
    "        (text_y := text_y_new)\n",
    "        if text_y - text_y_new > 0.75\n",
    "        else (text_y := text_y - 0.75)\n",
    "    )\n",
    "    plt.annotate(\n",
    "        gene,\n",
    "        xy=(gene_meta[x], gene_meta[y]),\n",
    "        xytext=(gene_meta[x] + 2, text_y_new),\n",
    "        arrowprops=dict(facecolor=\"black\", width=2, headwidth=5, headlength=5),\n",
    "        fontsize=12,\n",
    "    )\n",
    "\n",
    "text_y = 99\n",
    "for gene, gene_meta in top_down.iterrows():\n",
    "    text_y_new = gene_meta[y] + 0.75\n",
    "    text_y_new = (\n",
    "        (text_y := text_y_new)\n",
    "        if text_y - text_y_new > 0.75\n",
    "        else (text_y := text_y - 0.75)\n",
    "    )\n",
    "    plt.annotate(\n",
    "        gene,\n",
    "        xy=(gene_meta[x], gene_meta[y]),\n",
    "        xytext=(gene_meta[x] - 3, text_y_new),\n",
    "        arrowprops=dict(facecolor=\"black\", width=2, headwidth=5, headlength=5),\n",
    "        fontsize=12,\n",
    "    )\n",
    "\n",
    "\n",
    "plt.xlim(xmin, xmax)\n",
    "plt.ylim(ymin, ymax)\n",
    "plt.legend(loc=\"upper left\")\n",
    "plt.vlines(1, ymin=ymin, ymax=ymax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "plt.vlines(-1, ymin=ymin, ymax=ymax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "plt.hlines(1.96, xmin=xmin, xmax=xmax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "# plt.suptitle(\"Gene Occurrence of DEGs (Met over)\", y=1.03)\n",
    "plt.ylabel(\"Gene occurrence (Z-score normalised)\", fontsize=14)\n",
    "plt.xlabel(r\"$Log_{2}$ Fold Change\", fontsize=14)\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"4E_jointplot_diff_expr_met_over.pdf\"),\n",
    "    bbox_inches=\"tight\",\n",
    ")\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"4E_jointplot_diff_expr_met_over.svg\"),\n",
    "    bbox_inches=\"tight\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4F. Jointplot of z-scores vs. log2 fold changes of DMGs (met over)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90242956",
   "metadata": {},
   "outputs": [],
   "source": [
    "x, y, hue = [\n",
    "    \"diff_meth_promoters_meth.diff\",\n",
    "    \"mean_z_score\",\n",
    "    \"Statistical significance\",\n",
    "]\n",
    "meth_scatter_df = top_genes_df[[x, y]].dropna()\n",
    "\n",
    "meth_scatter_df[hue] = \"Significant\"\n",
    "meth_scatter_df.loc[meth_scatter_df[\"mean_z_score\"] < 1.96, hue] = \"Not Significant\"\n",
    "meth_scatter_df.loc[abs(meth_scatter_df[x]) < 10, hue] = \"Not Significant\"\n",
    "\n",
    "palette = {\"Significant\": \"#8B3A3A\", \"Not Significant\": \"#BFBFBF\"}\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6, 5), dpi=100)\n",
    "\n",
    "grid = sns.scatterplot(data=meth_scatter_df, x=x, y=y, hue=hue, palette=palette, ax=ax)\n",
    "\n",
    "top_up = (\n",
    "    meth_scatter_df.loc[\n",
    "        (meth_scatter_df[x] > 1) & (meth_scatter_df[hue] == \"Significant\")\n",
    "    ]\n",
    "    .sort_values(y, ascending=False)\n",
    "    .head(5)\n",
    ")\n",
    "top_down = (\n",
    "    meth_scatter_df.loc[\n",
    "        (meth_scatter_df[x] < -1) & (meth_scatter_df[hue] == \"Significant\")\n",
    "    ]\n",
    "    .sort_values(y, ascending=False)\n",
    "    .head(5)\n",
    ")\n",
    "\n",
    "text_y = 99\n",
    "for gene, gene_meta in top_up.iterrows():\n",
    "    text_y_new = gene_meta[y] + 0.75\n",
    "    text_y_new = (\n",
    "        (text_y := text_y_new)\n",
    "        if text_y - text_y_new > 0.75\n",
    "        else (text_y := text_y - 0.75)\n",
    "    )\n",
    "    plt.annotate(\n",
    "        gene,\n",
    "        xy=(gene_meta[x], gene_meta[y]),\n",
    "        xytext=(gene_meta[x] + 5, text_y_new),\n",
    "        arrowprops=dict(facecolor=\"black\", width=2, headwidth=5, headlength=5),\n",
    "        fontsize=12,\n",
    "    )\n",
    "\n",
    "text_y = 99\n",
    "for gene, gene_meta in top_down.iterrows():\n",
    "    text_y_new = gene_meta[y] + 0.75\n",
    "    text_y_new = (\n",
    "        (text_y := text_y_new)\n",
    "        if text_y - text_y_new > 0.75\n",
    "        else (text_y := text_y - 0.75)\n",
    "    )\n",
    "    plt.annotate(\n",
    "        gene,\n",
    "        xy=(gene_meta[x], gene_meta[y]),\n",
    "        xytext=(gene_meta[x] - 15, text_y_new),\n",
    "        arrowprops=dict(facecolor=\"black\", width=2, headwidth=5, headlength=5),\n",
    "        fontsize=12,\n",
    "    )\n",
    "\n",
    "plt.xlim(-40, 40)\n",
    "plt.ylim(ymin, ymax)\n",
    "plt.legend(loc=\"upper left\")\n",
    "plt.vlines(10, ymin=ymin, ymax=ymax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "plt.vlines(-10, ymin=ymin, ymax=ymax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "plt.hlines(1.96, xmin=xmin, xmax=xmax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "# plt.suptitle(\"Gene Occurrence of DMGs (Met over)\", y=1.03)\n",
    "plt.ylabel(\"Gene occurrence (Z-score normalised)\", fontsize=14)\n",
    "plt.xlabel(\"Mean Methylation Difference\", fontsize=14)\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"4F_jointplot_diff_meth_met_over.pdf\"),\n",
    "    bbox_inches=\"tight\",\n",
    ")\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"4F_jointplot_diff_meth_met_over.svg\"),\n",
    "    bbox_inches=\"tight\",\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "18029ba7",
   "metadata": {},
   "source": [
    "### 4G. Jointplot of z-scores vs. log2 fold changes of DEGs (met under)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43216c69",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_genes_df = pd.read_csv(\n",
    "    pcta_wcdt_root.joinpath(\"degss_genes_stats\")\n",
    "    .joinpath(\n",
    "        \"met_high_vs_met_low_degss_padj_0_05_down_0_0_\"\n",
    "        \"gsea_H_C1_C2_C3_C4_C5_C6_C7_C8_union\"\n",
    "    )\n",
    "    .joinpath(\"gene_occurrence_stats_bootstrap_z_score_summary_ann.csv\"),\n",
    "    index_col=0,\n",
    ")\n",
    "print(top_genes_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_genes_df.to_excel(writer_3, sheet_name=\"3D. Top genes stats (MET under)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a82f1ab7",
   "metadata": {},
   "outputs": [],
   "source": [
    "x, y, hue = [\"diff_expr_log2FoldChange\", \"mean_z_score\", \"Statistical significance\"]\n",
    "expr_scatter_df = top_genes_df[[x, y]].dropna()\n",
    "\n",
    "expr_scatter_df[hue] = \"Significant\"\n",
    "expr_scatter_df.loc[expr_scatter_df[\"mean_z_score\"] < 1.96, hue] = \"Not Significant\"\n",
    "expr_scatter_df.loc[abs(expr_scatter_df[x]) < 1, hue] = \"Not Significant\"\n",
    "\n",
    "palette = {\"Significant\": \"#8B3A3A\", \"Not Significant\": \"#BFBFBF\"}\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6, 5), dpi=100)\n",
    "\n",
    "grid = sns.scatterplot(data=expr_scatter_df, x=x, y=y, hue=hue, palette=palette, ax=ax)\n",
    "\n",
    "top_up = (\n",
    "    expr_scatter_df.loc[\n",
    "        (expr_scatter_df[x] > 1) & (expr_scatter_df[hue] == \"Significant\")\n",
    "    ]\n",
    "    .sort_values(y, ascending=False)\n",
    "    .head(5)\n",
    ")\n",
    "top_down = (\n",
    "    expr_scatter_df.loc[\n",
    "        (expr_scatter_df[x] < -1) & (expr_scatter_df[hue] == \"Significant\")\n",
    "    ]\n",
    "    .sort_values(y, ascending=False)\n",
    "    .head(5)\n",
    ")\n",
    "\n",
    "text_y = 99\n",
    "for gene, gene_meta in top_up.iterrows():\n",
    "    text_y_new = gene_meta[y] + 0.75\n",
    "    text_y_new = (\n",
    "        (text_y := text_y_new)\n",
    "        if text_y - text_y_new > 0.75\n",
    "        else (text_y := text_y - 0.75)\n",
    "    )\n",
    "    plt.annotate(\n",
    "        gene,\n",
    "        xy=(gene_meta[x], gene_meta[y]),\n",
    "        xytext=(gene_meta[x] + 2, text_y_new),\n",
    "        arrowprops=dict(facecolor=\"black\", width=2, headwidth=5, headlength=5),\n",
    "        fontsize=12,\n",
    "    )\n",
    "\n",
    "text_y = 99\n",
    "for gene, gene_meta in top_down.iterrows():\n",
    "    text_y_new = gene_meta[y] + 0.75\n",
    "    text_y_new = (\n",
    "        (text_y := text_y_new)\n",
    "        if text_y - text_y_new > 0.75\n",
    "        else (text_y := text_y - 0.75)\n",
    "    )\n",
    "    plt.annotate(\n",
    "        gene,\n",
    "        xy=(gene_meta[x], gene_meta[y]),\n",
    "        xytext=(gene_meta[x] - 3, text_y_new),\n",
    "        arrowprops=dict(facecolor=\"black\", width=2, headwidth=5, headlength=5),\n",
    "        fontsize=12,\n",
    "    )\n",
    "\n",
    "\n",
    "plt.xlim(xmin, xmax)\n",
    "plt.ylim(ymin, ymax)\n",
    "plt.legend(loc=\"upper left\")\n",
    "plt.vlines(1, ymin=ymin, ymax=ymax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "plt.vlines(-1, ymin=ymin, ymax=ymax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "plt.hlines(1.96, xmin=xmin, xmax=xmax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "# plt.suptitle(\"Gene Occurrence of DEGs (Met under)\", y=1.03)\n",
    "plt.ylabel(\"Gene occurrence (Z-score normalised)\", fontsize=14)\n",
    "plt.xlabel(r\"$Log_{2}$ Fold Change\", fontsize=14)\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"4G_jointplot_diff_expr_met_under.pdf\"),\n",
    "    bbox_inches=\"tight\",\n",
    ")\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"4G_jointplot_diff_expr_met_under.svg\"),\n",
    "    bbox_inches=\"tight\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4H. Jointplot of z-scores vs. log2 fold changes of DMGs (met under)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b91257d",
   "metadata": {},
   "outputs": [],
   "source": [
    "x, y, hue = [\n",
    "    \"diff_meth_promoters_meth.diff\",\n",
    "    \"mean_z_score\",\n",
    "    \"Statistical significance\",\n",
    "]\n",
    "meth_scatter_df = top_genes_df[[x, y]].dropna()\n",
    "\n",
    "meth_scatter_df[hue] = \"Significant\"\n",
    "meth_scatter_df.loc[meth_scatter_df[\"mean_z_score\"] < 1.96, hue] = \"Not Significant\"\n",
    "meth_scatter_df.loc[abs(meth_scatter_df[x]) < 5, hue] = \"Not Significant\"\n",
    "\n",
    "palette = {\"Significant\": \"#8B3A3A\", \"Not Significant\": \"#BFBFBF\"}\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6, 5), dpi=100)\n",
    "\n",
    "grid = sns.scatterplot(data=meth_scatter_df, x=x, y=y, hue=hue, palette=palette, ax=ax)\n",
    "\n",
    "top_up = (\n",
    "    meth_scatter_df.loc[\n",
    "        (meth_scatter_df[x] > 1) & (meth_scatter_df[hue] == \"Significant\")\n",
    "    ]\n",
    "    .sort_values(y, ascending=False)\n",
    "    .head(5)\n",
    ")\n",
    "top_down = (\n",
    "    meth_scatter_df.loc[\n",
    "        (meth_scatter_df[x] < -1) & (meth_scatter_df[hue] == \"Significant\")\n",
    "    ]\n",
    "    .sort_values(y, ascending=False)\n",
    "    .head(5)\n",
    ")\n",
    "\n",
    "text_y = 99\n",
    "for gene, gene_meta in top_up.iterrows():\n",
    "    text_y_new = gene_meta[y] + 0.75\n",
    "    text_y_new = (\n",
    "        (text_y := text_y_new)\n",
    "        if text_y - text_y_new > 0.75\n",
    "        else (text_y := text_y - 0.75)\n",
    "    )\n",
    "    plt.annotate(\n",
    "        gene,\n",
    "        xy=(gene_meta[x], gene_meta[y]),\n",
    "        xytext=(gene_meta[x] + 7, text_y_new),\n",
    "        arrowprops=dict(facecolor=\"black\", width=2, headwidth=5, headlength=5),\n",
    "        fontsize=12,\n",
    "    )\n",
    "\n",
    "text_y = 99\n",
    "for gene, gene_meta in top_down.iterrows():\n",
    "    text_y_new = gene_meta[y] + 1\n",
    "    text_y_new = (\n",
    "        (text_y := text_y_new)\n",
    "        if text_y - text_y_new > 1\n",
    "        else (text_y := text_y - 1)\n",
    "    )\n",
    "    plt.annotate(\n",
    "        gene,\n",
    "        xy=(gene_meta[x], gene_meta[y]),\n",
    "        xytext=(gene_meta[x] - 15, text_y_new),\n",
    "        arrowprops=dict(facecolor=\"black\", width=2, headwidth=5, headlength=5),\n",
    "        fontsize=12,\n",
    "    )\n",
    "\n",
    "plt.xlim(-40, 40)\n",
    "plt.ylim(ymin, ymax)\n",
    "plt.legend(loc=\"upper left\")\n",
    "plt.vlines(10, ymin=ymin, ymax=ymax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "plt.vlines(-10, ymin=ymin, ymax=ymax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "plt.hlines(1.96, xmin=xmin, xmax=xmax, colors=\"#8C8C8C\", linestyles=\"dashed\")\n",
    "# plt.suptitle(\"Gene Occurrence of DMGs (Met under)\", y=1.03)\n",
    "plt.ylabel(\"Gene occurrence (Z-score normalised)\", fontsize=14)\n",
    "plt.xlabel(\"Mean Methylation Difference\", fontsize=14)\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"4H_jointplot_diff_meth_met_under.pdf\"),\n",
    "    bbox_inches=\"tight\",\n",
    ")\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"4H_jointplot_diff_meth_met_under.svg\"),\n",
    "    bbox_inches=\"tight\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "writer_3.close()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "225ba399",
   "metadata": {},
   "source": [
    "## Figure 5: Functional profiling of PSMA protein community\n",
    "\n",
    "1. (**A**) Barplot & Stripplots of ORA results for category H (prim over)\n",
    "2. (**B**) Barplot & Stripplots of ORA results for category H (prim under)\n",
    "3. (**C**) Barplot & Stripplots of ORA results for category H (met over)\n",
    "4. (**D**) Barplot & Stripplots of ORA results for category H (met under)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "writer_4 = pd.ExcelWriter(pcta_wcdt_add_files_path / \"additional_file_4.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82101f22",
   "metadata": {},
   "outputs": [],
   "source": [
    "ytick_length = 36"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5A: Barplot & Stripplots of ORA results for category H (prim over)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clustering_root = (\n",
    "    pcta_wcdt_root.joinpath(\"degss_ppi_networks\")\n",
    "    .joinpath(\n",
    "        \"prim_high_vs_prim_low_degss_padj_0_05_up_0_0_\"\n",
    "        \"gsea_H_C1_C2_C3_C4_C5_C6_C7_C8_union\"\n",
    "    )\n",
    "    .joinpath(\"functional_network_at_500_score\")\n",
    "    .joinpath(\"clustering\")\n",
    "    .joinpath(\"p_1_0_q_0_5\")\n",
    ")\n",
    "\n",
    "clusters_df = pd.read_csv(\n",
    "    clustering_root.joinpath(\"genes_cluster_ids_metadata.csv\"), index_col=0\n",
    ").sort_values(\"cluster_id\", ascending=True)\n",
    "psma_cluster_id = clusters_df.loc[\"FOLH1\"][\"cluster_id\"]\n",
    "\n",
    "ora_h_file = (\n",
    "    clustering_root.joinpath(\"functional\")\n",
    "    .joinpath(\"MSIGDB\")\n",
    "    .joinpath(f\"{psma_cluster_id}_ora_H.csv\")\n",
    ")\n",
    "\n",
    "if ora_h_file.exists():\n",
    "    ora_h_df = (\n",
    "        pd.read_csv(ora_h_file, index_col=0)\n",
    "        .drop(columns=[\"ID\", \"Description\"])\n",
    "        .sort_values(\"Count\", ascending=False)\n",
    "    )\n",
    "\n",
    "ora_h_df.index = ora_h_df.index.str.replace(\"HALLMARK_\", \"\").str.replace(\"_\", \" \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_genes_df = pd.read_csv(\n",
    "    pcta_wcdt_root.joinpath(\"degss_genes_stats\")\n",
    "    .joinpath(\n",
    "        \"prim_high_vs_prim_low_degss_padj_0_05_up_0_0_\"\n",
    "        \"gsea_H_C1_C2_C3_C4_C5_C6_C7_C8_union\"\n",
    "    )\n",
    "    .joinpath(\"gene_occurrence_stats_bootstrap_z_score_summary_ann.csv\"),\n",
    "    index_col=0,\n",
    ")\n",
    "\n",
    "num_cols = [\"mean_z_score\", \"diff_expr_log2FoldChange\", \"diff_meth_promoters_logFC\"]\n",
    "data_df = (\n",
    "    ora_h_df[\"geneID\"]\n",
    "    .str.split(\"/\")\n",
    "    .explode()\n",
    "    .rename(\"gene\")\n",
    "    .to_frame()\n",
    "    .join(top_genes_df[num_cols], on=\"gene\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_df.to_excel(writer_4, sheet_name=\"4A. ORA H Genes (PRIM over)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 4, figsize=(24, 3), dpi=100)\n",
    "\n",
    "# 0. Count barplot\n",
    "sns.barplot(data=ora_h_df, x=\"Count\", y=ora_h_df.index, ax=ax[0], palette=palette_str)\n",
    "for container in ax[0].containers:\n",
    "    ax[0].bar_label(container, padding=5)\n",
    "\n",
    "# ax[0].set_ylabel(\"Gene Sets\", fontsize=14)\n",
    "ax[0].set_xlabel(\"Count\", fontsize=18)\n",
    "ax[0].set_yticklabels(\n",
    "    [\n",
    "        \" \" * (ytick_length - len(gene_set)) + gene_set\n",
    "        for gene_set in ora_h_df.index.tolist()\n",
    "    ],\n",
    "    fontsize=16,\n",
    ")\n",
    "ax[0].set_xlim([ax[0].get_xlim()[0] * 1.2, ax[0].get_xlim()[1] * 1.2])\n",
    "\n",
    "# 1. Mean Z-score stripplot\n",
    "data_df_plot = deepcopy(data_df.dropna(subset=[\"mean_z_score\"]))\n",
    "sns.boxplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"mean_z_score\",\n",
    "    ax=ax[1],\n",
    "    palette=palette_str,\n",
    "    boxprops=dict(alpha=0.5),\n",
    ")\n",
    "sns.stripplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"mean_z_score\",\n",
    "    hue=data_df_plot.index,\n",
    "    ax=ax[1],\n",
    "    palette=palette_str,\n",
    "    marker=\"o\",\n",
    "    linewidth=1,\n",
    "    jitter=0.2,\n",
    ")\n",
    "ax[1].get_legend().remove()\n",
    "ax[1].set_xlim([0, ax[1].get_xlim()[1] * 1.2])\n",
    "ax[1].set_yticklabels([])\n",
    "ax[1].set_xlabel(\"Mean Z-score\", fontsize=18)\n",
    "\n",
    "# 2. Diff. Expr. LFC\n",
    "data_df_plot = deepcopy(data_df.dropna(subset=[\"diff_expr_log2FoldChange\"]))\n",
    "sns.boxplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"diff_expr_log2FoldChange\",\n",
    "    ax=ax[2],\n",
    "    palette=palette_str,\n",
    "    boxprops=dict(alpha=0.5),\n",
    ")\n",
    "sns.stripplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"diff_expr_log2FoldChange\",\n",
    "    hue=data_df_plot.index,\n",
    "    ax=ax[2],\n",
    "    palette=palette_str,\n",
    "    marker=\"o\",\n",
    "    linewidth=1,\n",
    "    jitter=0.2,\n",
    ")\n",
    "ax[2].get_legend().remove()\n",
    "ax[2].set_xlim([ax[2].get_xlim()[0] * 1.2, ax[2].get_xlim()[1] * 1.2])\n",
    "ax[2].set_yticklabels([])\n",
    "ax[2].set_xlabel(\"Diff. Expr. $Log_2$ Fold Change\", fontsize=18)\n",
    "\n",
    "# 3. Diff. Meth. LFC\n",
    "data_df_plot = deepcopy(data_df.dropna(subset=[\"diff_meth_promoters_logFC\"]))\n",
    "sns.boxplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"diff_meth_promoters_logFC\",\n",
    "    ax=ax[3],\n",
    "    palette=palette_str,\n",
    "    boxprops=dict(alpha=0.5),\n",
    ")\n",
    "sns.stripplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"diff_meth_promoters_logFC\",\n",
    "    hue=data_df_plot.index,\n",
    "    ax=ax[3],\n",
    "    palette=palette_str,\n",
    "    marker=\"o\",\n",
    "    linewidth=1,\n",
    "    jitter=0.2,\n",
    ")\n",
    "ax[3].get_legend().remove()\n",
    "ax[3].set_xlim([ax[3].get_xlim()[0] * 1.2, ax[3].get_xlim()[1] * 1.2])\n",
    "ax[3].set_yticklabels([])\n",
    "ax[3].set_xlabel(\"Diff. Meth. $Log_2$ Fold Change\", fontsize=18)\n",
    "\n",
    "plt.subplots_adjust(wspace=0.08, hspace=0)\n",
    "# plt.suptitle(\n",
    "#     \"Over-representation analysis on Hallmark gene sets\\n(Prim. High vs Prim. low, over)\",\n",
    "#     fontsize=24,\n",
    "#     y=1.02,\n",
    "# )\n",
    "plt.tight_layout()\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"5A_ora_h_prim_over.pdf\"), bbox_inches=\"tight\")\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"5A_ora_h_prim_over.svg\"), bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_df[\n",
    "    (abs(data_df[\"diff_expr_log2FoldChange\"]) > 1)\n",
    "    | (abs(data_df[\"diff_meth_promoters_logFC\"]) > 1)\n",
    "].sort_values(\"diff_expr_log2FoldChange\", key=abs, ascending=False).drop_duplicates(\n",
    "    \"gene\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5B: Barplot & Stripplots of ORA results for category H (prim under)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clustering_root = (\n",
    "    pcta_wcdt_root.joinpath(\"degss_ppi_networks\")\n",
    "    .joinpath(\n",
    "        \"prim_high_vs_prim_low_degss_padj_0_05_down_0_0_\"\n",
    "        \"gsea_H_C1_C2_C3_C4_C5_C6_C7_C8_union\"\n",
    "    )\n",
    "    .joinpath(\"functional_network_at_500_score\")\n",
    "    .joinpath(\"clustering\")\n",
    "    .joinpath(\"p_1_0_q_0_5\")\n",
    ")\n",
    "\n",
    "clusters_df = pd.read_csv(\n",
    "    clustering_root.joinpath(\"genes_cluster_ids_metadata.csv\"), index_col=0\n",
    ").sort_values(\"cluster_id\", ascending=True)\n",
    "psma_cluster_id = clusters_df.loc[\"FOLH1\"][\"cluster_id\"]\n",
    "\n",
    "ora_h_file = (\n",
    "    clustering_root.joinpath(\"functional\")\n",
    "    .joinpath(\"MSIGDB\")\n",
    "    .joinpath(f\"{psma_cluster_id}_ora_H.csv\")\n",
    ")\n",
    "\n",
    "if ora_h_file.exists():\n",
    "    ora_h_df = (\n",
    "        pd.read_csv(ora_h_file, index_col=0)\n",
    "        .drop(columns=[\"ID\", \"Description\"])\n",
    "        .sort_values(\"Count\", ascending=False)\n",
    "    )\n",
    "\n",
    "ora_h_df.index = ora_h_df.index.str.replace(\"HALLMARK_\", \"\").str.replace(\"_\", \" \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_genes_df = pd.read_csv(\n",
    "    pcta_wcdt_root.joinpath(\"degss_genes_stats\")\n",
    "    .joinpath(\n",
    "        \"prim_high_vs_prim_low_degss_padj_0_05_down_0_0_\"\n",
    "        \"gsea_H_C1_C2_C3_C4_C5_C6_C7_C8_union\"\n",
    "    )\n",
    "    .joinpath(\"gene_occurrence_stats_bootstrap_z_score_summary_ann.csv\"),\n",
    "    index_col=0,\n",
    ")\n",
    "\n",
    "num_cols = [\"mean_z_score\", \"diff_expr_log2FoldChange\", \"diff_meth_promoters_logFC\"]\n",
    "data_df = (\n",
    "    ora_h_df[\"geneID\"]\n",
    "    .str.split(\"/\")\n",
    "    .explode()\n",
    "    .rename(\"gene\")\n",
    "    .to_frame()\n",
    "    .join(top_genes_df[num_cols], on=\"gene\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_df.to_excel(writer_4, sheet_name=\"4B. ORA H Genes (PRIM under)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 4, figsize=(24, 2.75), dpi=100)\n",
    "\n",
    "# 0. Count barplot\n",
    "sns.barplot(data=ora_h_df, x=\"Count\", y=ora_h_df.index, ax=ax[0], palette=palette_str)\n",
    "for container in ax[0].containers:\n",
    "    ax[0].bar_label(container, padding=5)\n",
    "\n",
    "# ax[0].set_ylabel(\"Gene Sets\", fontsize=14)\n",
    "ax[0].set_xlabel(\"Count\", fontsize=18)\n",
    "ax[0].set_yticklabels(\n",
    "    [\n",
    "        \" \" * (ytick_length - len(gene_set)) + gene_set\n",
    "        for gene_set in ora_h_df.index.tolist()\n",
    "    ],\n",
    "    fontsize=16,\n",
    ")\n",
    "ax[0].set_xlim([ax[0].get_xlim()[0] * 1.2, ax[0].get_xlim()[1] * 1.2])\n",
    "\n",
    "# 1. Mean Z-score stripplot\n",
    "data_df_plot = deepcopy(data_df.dropna(subset=[\"mean_z_score\"]))\n",
    "sns.boxplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"mean_z_score\",\n",
    "    ax=ax[1],\n",
    "    palette=palette_str,\n",
    "    boxprops=dict(alpha=0.5),\n",
    ")\n",
    "sns.stripplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"mean_z_score\",\n",
    "    hue=data_df_plot.index,\n",
    "    ax=ax[1],\n",
    "    palette=palette_str,\n",
    "    marker=\"o\",\n",
    "    linewidth=1,\n",
    "    jitter=0.2,\n",
    ")\n",
    "ax[1].get_legend().remove()\n",
    "ax[1].set_xlim([0, ax[1].get_xlim()[1] * 1.2])\n",
    "ax[1].set_yticklabels([])\n",
    "ax[1].set_xlabel(\"Mean Z-score\", fontsize=18)\n",
    "\n",
    "# 2. Diff. Expr. LFC\n",
    "data_df_plot = deepcopy(data_df.dropna(subset=[\"diff_expr_log2FoldChange\"]))\n",
    "sns.boxplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"diff_expr_log2FoldChange\",\n",
    "    ax=ax[2],\n",
    "    palette=palette_str,\n",
    "    boxprops=dict(alpha=0.5),\n",
    ")\n",
    "sns.stripplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"diff_expr_log2FoldChange\",\n",
    "    hue=data_df_plot.index,\n",
    "    ax=ax[2],\n",
    "    palette=palette_str,\n",
    "    marker=\"o\",\n",
    "    linewidth=1,\n",
    "    jitter=0.2,\n",
    ")\n",
    "ax[2].get_legend().remove()\n",
    "ax[2].set_xlim([ax[2].get_xlim()[0] * 1.2, ax[2].get_xlim()[1] * 1.2])\n",
    "ax[2].set_yticklabels([])\n",
    "ax[2].set_xlabel(\"Diff. Expr. $Log_2$ Fold Change\", fontsize=18)\n",
    "\n",
    "# 3. Diff. Meth. LFC\n",
    "data_df_plot = deepcopy(data_df.dropna(subset=[\"diff_meth_promoters_logFC\"]))\n",
    "sns.boxplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"diff_meth_promoters_logFC\",\n",
    "    ax=ax[3],\n",
    "    palette=palette_str,\n",
    "    boxprops=dict(alpha=0.5),\n",
    ")\n",
    "sns.stripplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"diff_meth_promoters_logFC\",\n",
    "    hue=data_df_plot.index,\n",
    "    ax=ax[3],\n",
    "    palette=palette_str,\n",
    "    marker=\"o\",\n",
    "    linewidth=1,\n",
    "    jitter=0.2,\n",
    ")\n",
    "ax[3].get_legend().remove()\n",
    "ax[3].set_xlim([ax[3].get_xlim()[0] * 1.2, ax[3].get_xlim()[1] * 1.2])\n",
    "ax[3].set_yticklabels([])\n",
    "ax[3].set_xlabel(\"Diff. Meth. $Log_2$ Fold Change\", fontsize=18)\n",
    "\n",
    "plt.subplots_adjust(wspace=0.08, hspace=0)\n",
    "# plt.suptitle(\n",
    "#     \"Over-representation analysis on Hallmark gene sets\\n(Prim. High vs Prim. low, under)\",\n",
    "#     fontsize=24,\n",
    "#     y=1.02,\n",
    "# )\n",
    "plt.tight_layout()\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"5B_ora_h_prim_under.pdf\"), bbox_inches=\"tight\"\n",
    ")\n",
    "plt.savefig(\n",
    "    pcta_wcdt_figs_path.joinpath(\"5B_ora_h_prim_under.svg\"), bbox_inches=\"tight\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_df[\n",
    "    (abs(data_df[\"diff_expr_log2FoldChange\"]) > 1)\n",
    "    | (abs(data_df[\"diff_meth_promoters_logFC\"]) > 1)\n",
    "].sort_values(\"diff_expr_log2FoldChange\", key=abs, ascending=False).drop_duplicates(\n",
    "    \"gene\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5C: Barplot & Stripplots of ORA results for category H (met over)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clustering_root = (\n",
    "    pcta_wcdt_root.joinpath(\"degss_ppi_networks\")\n",
    "    .joinpath(\n",
    "        \"met_high_vs_met_low_degss_padj_0_05_up_0_0_\"\n",
    "        \"gsea_H_C1_C2_C3_C4_C5_C6_C7_C8_union\"\n",
    "    )\n",
    "    .joinpath(\"functional_network_at_500_score\")\n",
    "    .joinpath(\"clustering\")\n",
    "    .joinpath(\"p_1_0_q_0_5\")\n",
    ")\n",
    "\n",
    "clusters_df = pd.read_csv(\n",
    "    clustering_root.joinpath(\"genes_cluster_ids_metadata.csv\"), index_col=0\n",
    ").sort_values(\"cluster_id\", ascending=True)\n",
    "psma_cluster_id = clusters_df.loc[\"FOLH1\"][\"cluster_id\"]\n",
    "\n",
    "ora_h_file = (\n",
    "    clustering_root.joinpath(\"functional\")\n",
    "    .joinpath(\"MSIGDB\")\n",
    "    .joinpath(f\"{psma_cluster_id}_ora_H.csv\")\n",
    ")\n",
    "\n",
    "if ora_h_file.exists():\n",
    "    ora_h_df = (\n",
    "        pd.read_csv(ora_h_file, index_col=0)\n",
    "        .drop(columns=[\"ID\", \"Description\"])\n",
    "        .sort_values(\"Count\", ascending=False)\n",
    "    )\n",
    "\n",
    "ora_h_df.index = ora_h_df.index.str.replace(\"HALLMARK_\", \"\").str.replace(\"_\", \" \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_genes_df = pd.read_csv(\n",
    "    pcta_wcdt_root.joinpath(\"degss_genes_stats\")\n",
    "    .joinpath(\n",
    "        \"met_high_vs_met_low_degss_padj_0_05_up_0_0_\"\n",
    "        \"gsea_H_C1_C2_C3_C4_C5_C6_C7_C8_union\"\n",
    "    )\n",
    "    .joinpath(\"gene_occurrence_stats_bootstrap_z_score_summary_ann.csv\"),\n",
    "    index_col=0,\n",
    ")\n",
    "\n",
    "num_cols = [\"mean_z_score\", \"diff_expr_log2FoldChange\", \"diff_meth_promoters_meth.diff\"]\n",
    "data_df = (\n",
    "    ora_h_df[\"geneID\"]\n",
    "    .str.split(\"/\")\n",
    "    .explode()\n",
    "    .rename(\"gene\")\n",
    "    .to_frame()\n",
    "    .join(top_genes_df[num_cols], on=\"gene\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_df.to_excel(writer_4, sheet_name=\"4C. ORA H Genes (MET over)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 4, figsize=(24, 1.75), dpi=100)\n",
    "\n",
    "# 0. Count barplot\n",
    "sns.barplot(data=ora_h_df, x=\"Count\", y=ora_h_df.index, ax=ax[0], palette=palette_str)\n",
    "for container in ax[0].containers:\n",
    "    ax[0].bar_label(container, padding=5)\n",
    "\n",
    "ax[0].set_xlabel(\"Count\", fontsize=18)\n",
    "ax[0].set_yticklabels(\n",
    "    [\n",
    "        \" \" * (ytick_length - len(gene_set)) + gene_set\n",
    "        for gene_set in ora_h_df.index.tolist()\n",
    "    ],\n",
    "    fontsize=16,\n",
    ")\n",
    "ax[0].set_xlim([ax[0].get_xlim()[0] * 1.2, ax[0].get_xlim()[1] * 1.2])\n",
    "\n",
    "# 1. Mean Z-score stripplot\n",
    "data_df_plot = deepcopy(data_df.dropna(subset=[\"mean_z_score\"]))\n",
    "sns.boxplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"mean_z_score\",\n",
    "    ax=ax[1],\n",
    "    palette=palette_str,\n",
    "    boxprops=dict(alpha=0.5),\n",
    ")\n",
    "sns.stripplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"mean_z_score\",\n",
    "    hue=data_df_plot.index,\n",
    "    ax=ax[1],\n",
    "    palette=palette_str,\n",
    "    marker=\"o\",\n",
    "    linewidth=1,\n",
    "    jitter=0.2,\n",
    ")\n",
    "ax[1].get_legend().remove()\n",
    "ax[1].set_xlim([0, ax[1].get_xlim()[1] * 1.2])\n",
    "ax[1].set_yticklabels([])\n",
    "ax[1].set_xlabel(\"Mean Z-score\", fontsize=18)\n",
    "\n",
    "# 2. Diff. Expr. LFC\n",
    "data_df_plot = deepcopy(data_df.dropna(subset=[\"diff_expr_log2FoldChange\"]))\n",
    "sns.boxplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"diff_expr_log2FoldChange\",\n",
    "    ax=ax[2],\n",
    "    palette=palette_str,\n",
    "    boxprops=dict(alpha=0.5),\n",
    ")\n",
    "sns.stripplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"diff_expr_log2FoldChange\",\n",
    "    hue=data_df_plot.index,\n",
    "    ax=ax[2],\n",
    "    palette=palette_str,\n",
    "    marker=\"o\",\n",
    "    linewidth=1,\n",
    "    jitter=0.2,\n",
    ")\n",
    "ax[2].get_legend().remove()\n",
    "ax[2].set_xlim([ax[2].get_xlim()[0] * 1.2, ax[2].get_xlim()[1] * 1.2])\n",
    "ax[2].set_yticklabels([])\n",
    "ax[2].set_xlabel(\"Diff. Expr. $Log_2$ Fold Change\", fontsize=18)\n",
    "\n",
    "# 3. Diff. Meth. LFC\n",
    "data_df_plot = deepcopy(data_df.dropna(subset=[\"diff_meth_promoters_meth.diff\"]))\n",
    "sns.boxplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"diff_meth_promoters_meth.diff\",\n",
    "    ax=ax[3],\n",
    "    palette=palette_str,\n",
    "    boxprops=dict(alpha=0.5),\n",
    ")\n",
    "sns.stripplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"diff_meth_promoters_meth.diff\",\n",
    "    hue=data_df_plot.index,\n",
    "    ax=ax[3],\n",
    "    palette=palette_str,\n",
    "    marker=\"o\",\n",
    "    linewidth=1,\n",
    "    jitter=0.2,\n",
    ")\n",
    "ax[3].get_legend().remove()\n",
    "ax[3].set_xlim([ax[3].get_xlim()[0] * 1.2, ax[3].get_xlim()[1] * 1.2])\n",
    "ax[3].set_yticklabels([])\n",
    "ax[3].set_xlabel(\"Diff. Meth. Difference\", fontsize=18)\n",
    "\n",
    "plt.subplots_adjust(wspace=0.08, hspace=0)\n",
    "# plt.suptitle(\n",
    "#     \"Over-representation analysis on Hallmark gene sets\\n(Met. High vs Met. low, over)\",\n",
    "#     fontsize=24,\n",
    "#     y=1.02,\n",
    "# )\n",
    "plt.tight_layout()\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"5C_ora_h_met_over.pdf\"), bbox_inches=\"tight\")\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"5C_ora_h_met_over.svg\"), bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_df[\n",
    "    (abs(data_df[\"diff_expr_log2FoldChange\"]) > 1)\n",
    "    | (abs(data_df[\"diff_meth_promoters_meth.diff\"]) > 10)\n",
    "].sort_values(\"diff_expr_log2FoldChange\", key=abs, ascending=False).drop_duplicates(\n",
    "    \"gene\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5D: Barplot & Stripplots of ORA results for category H (met under)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clustering_root = (\n",
    "    pcta_wcdt_root.joinpath(\"degss_ppi_networks\")\n",
    "    .joinpath(\n",
    "        \"met_high_vs_met_low_degss_padj_0_05_down_0_0_\"\n",
    "        \"gsea_H_C1_C2_C3_C4_C5_C6_C7_C8_union\"\n",
    "    )\n",
    "    .joinpath(\"functional_network_at_500_score\")\n",
    "    .joinpath(\"clustering\")\n",
    "    .joinpath(\"p_1_0_q_0_5\")\n",
    ")\n",
    "\n",
    "clusters_df = pd.read_csv(\n",
    "    clustering_root.joinpath(\"genes_cluster_ids_metadata.csv\"), index_col=0\n",
    ").sort_values(\"cluster_id\", ascending=True)\n",
    "psma_cluster_id = clusters_df.loc[\"FOLH1\"][\"cluster_id\"]\n",
    "\n",
    "ora_h_file = (\n",
    "    clustering_root.joinpath(\"functional\")\n",
    "    .joinpath(\"MSIGDB\")\n",
    "    .joinpath(f\"{psma_cluster_id}_ora_H.csv\")\n",
    ")\n",
    "\n",
    "if ora_h_file.exists():\n",
    "    ora_h_df = (\n",
    "        pd.read_csv(ora_h_file, index_col=0)\n",
    "        .drop(columns=[\"ID\", \"Description\"])\n",
    "        .sort_values(\"Count\", ascending=False)\n",
    "    )\n",
    "\n",
    "ora_h_df.index = ora_h_df.index.str.replace(\"HALLMARK_\", \"\").str.replace(\"_\", \" \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_genes_df = pd.read_csv(\n",
    "    pcta_wcdt_root.joinpath(\"degss_genes_stats\")\n",
    "    .joinpath(\n",
    "        \"met_high_vs_met_low_degss_padj_0_05_down_0_0_\"\n",
    "        \"gsea_H_C1_C2_C3_C4_C5_C6_C7_C8_union\"\n",
    "    )\n",
    "    .joinpath(\"gene_occurrence_stats_bootstrap_z_score_summary_ann.csv\"),\n",
    "    index_col=0,\n",
    ")\n",
    "\n",
    "num_cols = [\"mean_z_score\", \"diff_expr_log2FoldChange\", \"diff_meth_promoters_meth.diff\"]\n",
    "data_df = (\n",
    "    ora_h_df[\"geneID\"]\n",
    "    .str.split(\"/\")\n",
    "    .explode()\n",
    "    .rename(\"gene\")\n",
    "    .to_frame()\n",
    "    .join(top_genes_df[num_cols], on=\"gene\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_df.to_excel(writer_4, sheet_name=\"4D. ORA H Genes (MET under)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 4, figsize=(24, 4), dpi=100)\n",
    "\n",
    "# 0. Count barplot\n",
    "sns.barplot(data=ora_h_df, x=\"Count\", y=ora_h_df.index, ax=ax[0], palette=palette_str)\n",
    "for container in ax[0].containers:\n",
    "    ax[0].bar_label(container, padding=5)\n",
    "\n",
    "ax[0].set_xlabel(\"Count\", fontsize=18)\n",
    "ax[0].set_yticklabels(\n",
    "    [\n",
    "        \" \" * (ytick_length - len(gene_set)) + gene_set\n",
    "        for gene_set in ora_h_df.index.tolist()\n",
    "    ],\n",
    "    fontsize=16,\n",
    ")\n",
    "ax[0].set_xlim([ax[0].get_xlim()[0] * 1.2, ax[0].get_xlim()[1] * 1.2])\n",
    "\n",
    "# 1. Mean Z-score stripplot\n",
    "data_df_plot = deepcopy(data_df.dropna(subset=[\"mean_z_score\"]))\n",
    "sns.boxplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"mean_z_score\",\n",
    "    ax=ax[1],\n",
    "    palette=palette_str,\n",
    "    boxprops=dict(alpha=0.5),\n",
    ")\n",
    "sns.stripplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"mean_z_score\",\n",
    "    hue=data_df_plot.index,\n",
    "    ax=ax[1],\n",
    "    palette=palette_str,\n",
    "    marker=\"o\",\n",
    "    linewidth=1,\n",
    "    jitter=0.2,\n",
    ")\n",
    "ax[1].get_legend().remove()\n",
    "ax[1].set_xlim([0, ax[1].get_xlim()[1] * 1.2])\n",
    "ax[1].set_yticklabels([])\n",
    "ax[1].set_xlabel(\"Mean Z-score\", fontsize=18)\n",
    "\n",
    "# 2. Diff. Expr. LFC\n",
    "data_df_plot = deepcopy(data_df.dropna(subset=[\"diff_expr_log2FoldChange\"]))\n",
    "sns.boxplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"diff_expr_log2FoldChange\",\n",
    "    ax=ax[2],\n",
    "    palette=palette_str,\n",
    "    boxprops=dict(alpha=0.5),\n",
    ")\n",
    "sns.stripplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"diff_expr_log2FoldChange\",\n",
    "    hue=data_df_plot.index,\n",
    "    ax=ax[2],\n",
    "    palette=palette_str,\n",
    "    marker=\"o\",\n",
    "    linewidth=1,\n",
    "    jitter=0.2,\n",
    ")\n",
    "ax[2].get_legend().remove()\n",
    "ax[2].set_xlim([ax[2].get_xlim()[0] * 1.2, ax[2].get_xlim()[1] * 1.2])\n",
    "ax[2].set_yticklabels([])\n",
    "ax[2].set_xlabel(\"Diff. Expr. $Log_2$ Fold Change\", fontsize=18)\n",
    "\n",
    "# 3. Diff. Meth. LFC\n",
    "data_df_plot = deepcopy(data_df.dropna(subset=[\"diff_meth_promoters_meth.diff\"]))\n",
    "sns.boxplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"diff_meth_promoters_meth.diff\",\n",
    "    ax=ax[3],\n",
    "    palette=palette_str,\n",
    "    boxprops=dict(alpha=0.5),\n",
    ")\n",
    "sns.stripplot(\n",
    "    data=data_df_plot,\n",
    "    y=data_df_plot.index,\n",
    "    x=\"diff_meth_promoters_meth.diff\",\n",
    "    hue=data_df_plot.index,\n",
    "    ax=ax[3],\n",
    "    palette=palette_str,\n",
    "    marker=\"o\",\n",
    "    linewidth=1,\n",
    "    jitter=0.2,\n",
    ")\n",
    "ax[3].get_legend().remove()\n",
    "ax[3].set_xlim([ax[3].get_xlim()[0] * 1.2, ax[3].get_xlim()[1] * 1.2])\n",
    "ax[3].set_yticklabels([])\n",
    "ax[3].set_xlabel(\"Diff. Meth. Difference\", fontsize=18)\n",
    "\n",
    "plt.subplots_adjust(wspace=0.08, hspace=0)\n",
    "# plt.suptitle(\n",
    "#     \"Over-representation analysis on Hallmark gene sets\\n(Met. High vs Met. low, under)\",\n",
    "#     fontsize=24,\n",
    "#     y=1.02,\n",
    "# )\n",
    "plt.tight_layout()\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"5D_ora_h_met_under.pdf\"), bbox_inches=\"tight\")\n",
    "plt.savefig(pcta_wcdt_figs_path.joinpath(\"5D_ora_h_met_under.svg\"), bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_df[\n",
    "    (abs(data_df[\"diff_expr_log2FoldChange\"]) > 1)\n",
    "    | (abs(data_df[\"diff_meth_promoters_meth.diff\"]) > 10)\n",
    "].sort_values(\"diff_expr_log2FoldChange\", key=abs, ascending=False).drop_duplicates(\n",
    "    \"gene\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "writer_4.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78c7629a",
   "metadata": {},
   "source": [
    "## Additional information and figures"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e253ab1",
   "metadata": {},
   "source": [
    "### Table 1: Enriched gene sets from different methods"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ba1f685",
   "metadata": {},
   "source": [
    "#### Primary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bec8eb08",
   "metadata": {},
   "outputs": [],
   "source": [
    "prim_dgsva_files = sorted(\n",
    "    list(\n",
    "        pcta_wcdt_root.joinpath(\"diff_gsva\").glob(\n",
    "            \"**/*prim_high_vs_prim_low_top_table_padj_0_05_up_0_0.csv\"\n",
    "        )\n",
    "    )\n",
    "    + list(\n",
    "        pcta_wcdt_root.joinpath(\"diff_gsva\").glob(\n",
    "            \"**/*prim_high_vs_prim_low_top_table_padj_0_05_down_0_0.csv\"\n",
    "        )\n",
    "    )\n",
    ")\n",
    "\n",
    "prim_dgsva_counts = dict()\n",
    "for dgsva_file in prim_dgsva_files:\n",
    "    msigdb_cat = dgsva_file.parent.stem\n",
    "    lfc_level = (\n",
    "        \"under-enriched\"\n",
    "        if dgsva_file.stem.split(\"_\")[-3] == \"down\"\n",
    "        else \"over-enriched\"\n",
    "    )\n",
    "    prim_dgsva_counts[(msigdb_cat, lfc_level)] = len(\n",
    "        pd.read_csv(dgsva_file, index_col=0)\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5c10410",
   "metadata": {},
   "outputs": [],
   "source": [
    "prim_gsea_de = sorted(\n",
    "    pcta_wcdt_root.joinpath(\"functional\")\n",
    "    .joinpath(\"MSIGDB\")\n",
    "    .glob(\"*_prim_high_vs_prim_low_gsea_*.csv\")\n",
    ")\n",
    "\n",
    "prim_gsea_de_counts = dict()\n",
    "for gsea_file in prim_gsea_de:\n",
    "    msigdb_cat = gsea_file.stem.split(\"_\")[-1]\n",
    "    df = pd.read_csv(gsea_file, index_col=0)\n",
    "    prim_gsea_de_counts[(msigdb_cat, \"over-enriched\")] = len(\n",
    "        df[df[\"enrichmentScore\"] > 0]\n",
    "    )\n",
    "    prim_gsea_de_counts[(msigdb_cat, \"under-enriched\")] = len(\n",
    "        df[df[\"enrichmentScore\"] < 0]\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "805a1c83",
   "metadata": {},
   "outputs": [],
   "source": [
    "prim_gsea_dm = sorted(\n",
    "    tcga_root.joinpath(\"functional\")\n",
    "    .joinpath(\"MSIGDB\")\n",
    "    .glob(\"*_prim_high_vs_prim_low_wrt_mean_diff_0_0_hg38_genes_promoters_gsea_*.csv\")\n",
    ")\n",
    "\n",
    "prim_gsea_dm_counts = dict()\n",
    "for gsea_file in prim_gsea_dm:\n",
    "    msigdb_cat = gsea_file.stem.split(\"_\")[-1]\n",
    "    df = pd.read_csv(gsea_file, index_col=0)\n",
    "    prim_gsea_dm_counts[(msigdb_cat, \"over-enriched\")] = len(\n",
    "        df[df[\"enrichmentScore\"] > 0]\n",
    "    )\n",
    "    prim_gsea_dm_counts[(msigdb_cat, \"under-enriched\")] = len(\n",
    "        df[df[\"enrichmentScore\"] < 0]\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e17c7ea",
   "metadata": {},
   "source": [
    "#### Metastatic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aeb68853",
   "metadata": {},
   "outputs": [],
   "source": [
    "met_dgsva_files = sorted(\n",
    "    list(\n",
    "        pcta_wcdt_root.joinpath(\"diff_gsva\").glob(\n",
    "            \"**/*met_high_vs_met_low_top_table_padj_0_05_up_0_0.csv\"\n",
    "        )\n",
    "    )\n",
    "    + list(\n",
    "        pcta_wcdt_root.joinpath(\"diff_gsva\").glob(\n",
    "            \"**/*met_high_vs_met_low_top_table_padj_0_05_down_0_0.csv\"\n",
    "        )\n",
    "    )\n",
    ")\n",
    "\n",
    "met_dgsva_counts = dict()\n",
    "for dgsva_file in met_dgsva_files:\n",
    "    msigdb_cat = dgsva_file.parent.stem\n",
    "    lfc_level = (\n",
    "        \"under-enriched\"\n",
    "        if dgsva_file.stem.split(\"_\")[-3] == \"down\"\n",
    "        else \"over-enriched\"\n",
    "    )\n",
    "    met_dgsva_counts[(msigdb_cat, lfc_level)] = len(\n",
    "        pd.read_csv(dgsva_file, index_col=0)\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d4b7b22",
   "metadata": {},
   "outputs": [],
   "source": [
    "met_gsea_de = sorted(\n",
    "    pcta_wcdt_root.joinpath(\"functional\")\n",
    "    .joinpath(\"MSIGDB\")\n",
    "    .glob(\"*_met_high_vs_met_low_gsea_*.csv\")\n",
    ")\n",
    "\n",
    "met_gsea_de_counts = dict()\n",
    "for gsea_file in met_gsea_de:\n",
    "    msigdb_cat = gsea_file.stem.split(\"_\")[-1]\n",
    "    df = pd.read_csv(gsea_file, index_col=0)\n",
    "    met_gsea_de_counts[(msigdb_cat, \"over-enriched\")] = len(\n",
    "        df[df[\"enrichmentScore\"] > 0]\n",
    "    )\n",
    "    met_gsea_de_counts[(msigdb_cat, \"under-enriched\")] = len(\n",
    "        df[df[\"enrichmentScore\"] < 0]\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98dc703e",
   "metadata": {},
   "outputs": [],
   "source": [
    "met_gsea_dm = sorted(\n",
    "    wcdt_root.joinpath(\"methylkit\")\n",
    "    .joinpath(\"functional\")\n",
    "    .joinpath(\"MSIGDB\")\n",
    "    .glob(\"*_met_high_vs_met_low_ann_genes_hg38_genes_promoters_gsea_*.csv\")\n",
    ")\n",
    "\n",
    "met_gsea_dm_counts = dict()\n",
    "for gsea_file in met_gsea_dm:\n",
    "    msigdb_cat = gsea_file.stem.split(\"_\")[-1]\n",
    "    df = pd.read_csv(gsea_file, index_col=0)\n",
    "    met_gsea_dm_counts[(msigdb_cat, \"over-enriched\")] = len(\n",
    "        df[df[\"enrichmentScore\"] > 0]\n",
    "    )\n",
    "    met_gsea_dm_counts[(msigdb_cat, \"under-enriched\")] = len(\n",
    "        df[df[\"enrichmentScore\"] < 0]\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "872daaf8",
   "metadata": {},
   "source": [
    "#### Build table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecbfafde",
   "metadata": {},
   "outputs": [],
   "source": [
    "table_1_df = (\n",
    "    pd.DataFrame(\n",
    "        {\n",
    "            (\"Prim high/low\", \"DGSVA\"): prim_dgsva_counts,\n",
    "            (\"Prim high/low\", \"GSEA DE\"): prim_gsea_de_counts,\n",
    "            (\"Prim high/low\", \"GSEA DM\"): prim_gsea_dm_counts,\n",
    "            (\"Met high/low\", \"DGSVA\"): met_dgsva_counts,\n",
    "            (\"Met high/low\", \"GSEA DE\"): met_gsea_de_counts,\n",
    "            (\"Met high/low\", \"GSEA DM\"): met_gsea_dm_counts,\n",
    "        }\n",
    "    )\n",
    "    .fillna(0)\n",
    "    .astype(int)\n",
    ")\n",
    "table_1_df.index.set_names([\"MSigDB Category\", \"Enrichment Level\"], inplace=True)\n",
    "table_1_df.to_csv(pcta_wcdt_add_files_path / \"table_1.csv\")\n",
    "table_1_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3ee77ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "table_1_df.to_latex(\n",
    "    pcta_wcdt_add_files_path / \"table_1.tex\",\n",
    "    bold_rows=True,\n",
    "    caption=\"Number of (differentially) enriched gene sets for each MSigDB category and enrichment method.\",\n",
    "    label=\"table1\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30e6317d",
   "metadata": {},
   "source": [
    "### Table 2: Cluster sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "288ad72f",
   "metadata": {},
   "outputs": [],
   "source": [
    "writer_3 = pd.ExcelWriter(pcta_wcdt_add_files_path / \"additional_file_3.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a598ac72",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_sizes = dict()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efc74c5d",
   "metadata": {},
   "source": [
    "#### Prim over"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0330bb50",
   "metadata": {},
   "outputs": [],
   "source": [
    "clustering_root = (\n",
    "    pcta_wcdt_root.joinpath(\"degss_ppi_networks\")\n",
    "    .joinpath(\n",
    "        \"prim_high_vs_prim_low_degss_padj_0_05_up_0_0_\"\n",
    "        \"gsea_H_C1_C2_C3_C4_C5_C6_C7_C8_union\"\n",
    "    )\n",
    "    .joinpath(\"functional_network_at_500_score\")\n",
    "    .joinpath(\"clustering\")\n",
    "    .joinpath(\"p_1_0_q_0_5\")\n",
    ")\n",
    "\n",
    "clusters_df = pd.read_csv(\n",
    "    clustering_root.joinpath(\"genes_cluster_ids_metadata.csv\"), index_col=0\n",
    ").sort_values(\"cluster_id\", ascending=True)\n",
    "psma_cluster_id = clusters_df.loc[\"FOLH1\"][\"cluster_id\"]\n",
    "print(\"Cluster ID FOLH1 is in:\", psma_cluster_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7abb462d",
   "metadata": {},
   "outputs": [],
   "source": [
    "clusters_df.to_excel(writer_3, sheet_name=\"3E. PPI Clustering (PRIM over)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b68124f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_sizes[\"prim_over\"] = (\n",
    "    clusters_df.groupby(\"cluster_id\")[\"FOLH1_distance\"].count().to_dict()\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4013815c",
   "metadata": {},
   "source": [
    "#### Prim under"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f81dd6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "clustering_root = (\n",
    "    pcta_wcdt_root.joinpath(\"degss_ppi_networks\")\n",
    "    .joinpath(\n",
    "        \"prim_high_vs_prim_low_degss_padj_0_05_down_0_0_\"\n",
    "        \"gsea_H_C1_C2_C3_C4_C5_C6_C7_C8_union\"\n",
    "    )\n",
    "    .joinpath(\"functional_network_at_500_score\")\n",
    "    .joinpath(\"clustering\")\n",
    "    .joinpath(\"p_1_0_q_0_5\")\n",
    ")\n",
    "\n",
    "clusters_df = pd.read_csv(\n",
    "    clustering_root.joinpath(\"genes_cluster_ids_metadata.csv\"), index_col=0\n",
    ").sort_values(\"cluster_id\", ascending=True)\n",
    "psma_cluster_id = clusters_df.loc[\"FOLH1\"][\"cluster_id\"]\n",
    "print(\"Cluster ID FOLH1 is in:\", psma_cluster_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "556a3cf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "clusters_df.to_excel(writer_3, sheet_name=\"3F. PPI Clustering (PRIM under)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "868e3c2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_sizes[\"prim_under\"] = (\n",
    "    clusters_df.groupby(\"cluster_id\")[\"FOLH1_distance\"].count().to_dict()\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f552035c",
   "metadata": {},
   "source": [
    "#### Met over"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d479326a",
   "metadata": {},
   "outputs": [],
   "source": [
    "clustering_root = (\n",
    "    pcta_wcdt_root.joinpath(\"degss_ppi_networks\")\n",
    "    .joinpath(\n",
    "        \"met_high_vs_met_low_degss_padj_0_05_up_0_0_\"\n",
    "        \"gsea_H_C1_C2_C3_C4_C5_C6_C7_C8_union\"\n",
    "    )\n",
    "    .joinpath(\"functional_network_at_500_score\")\n",
    "    .joinpath(\"clustering\")\n",
    "    .joinpath(\"p_1_0_q_0_5\")\n",
    ")\n",
    "\n",
    "clusters_df = pd.read_csv(\n",
    "    clustering_root.joinpath(\"genes_cluster_ids_metadata.csv\"), index_col=0\n",
    ").sort_values(\"cluster_id\", ascending=True)\n",
    "psma_cluster_id = clusters_df.loc[\"FOLH1\"][\"cluster_id\"]\n",
    "print(\"Cluster ID FOLH1 is in:\", psma_cluster_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92299f21",
   "metadata": {},
   "outputs": [],
   "source": [
    "clusters_df.to_excel(writer_3, sheet_name=\"3G. PPI Clustering (MET over)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d823dbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_sizes[\"met_over\"] = (\n",
    "    clusters_df.groupby(\"cluster_id\")[\"FOLH1_distance\"].count().to_dict()\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f51e5708",
   "metadata": {},
   "source": [
    "#### Met under"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ed319e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "clustering_root = (\n",
    "    pcta_wcdt_root.joinpath(\"degss_ppi_networks\")\n",
    "    .joinpath(\n",
    "        \"met_high_vs_met_low_degss_padj_0_05_down_0_0_\"\n",
    "        \"gsea_H_C1_C2_C3_C4_C5_C6_C7_C8_union\"\n",
    "    )\n",
    "    .joinpath(\"functional_network_at_500_score\")\n",
    "    .joinpath(\"clustering\")\n",
    "    .joinpath(\"p_1_0_q_0_5\")\n",
    ")\n",
    "\n",
    "clusters_df = pd.read_csv(\n",
    "    clustering_root.joinpath(\"genes_cluster_ids_metadata.csv\"), index_col=0\n",
    ").sort_values(\"cluster_id\", ascending=True)\n",
    "psma_cluster_id = clusters_df.loc[\"FOLH1\"][\"cluster_id\"]\n",
    "print(\"Cluster ID FOLH1 is in:\", psma_cluster_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a1fa396",
   "metadata": {},
   "outputs": [],
   "source": [
    "clusters_df.to_excel(writer_3, sheet_name=\"3H. PPI Clustering (MET under)\")\n",
    "writer_3.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93942eaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_sizes[\"met_under\"] = (\n",
    "    clusters_df.groupby(\"cluster_id\")[\"FOLH1_distance\"].count().to_dict()\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "847a66a8",
   "metadata": {},
   "source": [
    "#### Build table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88393b79",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_sizes_df = (\n",
    "    pd.DataFrame(cluster_sizes).fillna(0).astype(int).rename_axis(\"Cluster ID\")\n",
    ")\n",
    "cluster_sizes_df.to_csv(pcta_wcdt_add_files_path / \"table_2.csv\")\n",
    "cluster_sizes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67fbda7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_sizes_df.to_latex(\n",
    "    pcta_wcdt_add_files_path / \"table_2.tex\",\n",
    "    bold_rows=True,\n",
    "    caption=\"Size of clusters after ensemble clustering using Node2Vec embeddings.\",\n",
    "    label=\"table2\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8437768",
   "metadata": {},
   "source": [
    "### How many genes have both hypo- and hyper-methylated CpGs?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5050a85",
   "metadata": {},
   "source": [
    "#### Primary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94af1e2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_meth_df = (\n",
    "    pd.read_csv(\n",
    "        (\n",
    "            tcga_root.joinpath(\"minfi\").joinpath(\n",
    "                \"sample_type_prim_FOLH1_level_prim_high+prim_low__\"\n",
    "                \"diff_meth_probes_noob_quantile_top_table_prim_high_vs_prim_low_\"\n",
    "                \"wrt_mean_diff_0_0_hg38_genes_promoters.csv\"\n",
    "            )\n",
    "        ),\n",
    "        index_col=0,\n",
    "    )\n",
    "    .replace(\"NA_character_\", np.nan)\n",
    "    .dropna(subset=[\"annot.symbol\"])\n",
    ")\n",
    "\n",
    "hypo_hyper_meth_counts = (\n",
    "    diff_meth_df.groupby(\"annot.symbol\")[\"logFC\"]\n",
    "    .apply(lambda x: pd.Series([(x < 0).sum(), (x >= 0).sum()]))\n",
    "    .unstack()\n",
    "    .rename(columns={0: \"Hypo-methylated\", 1: \"Hyper-methylated\"})\n",
    ")\n",
    "\n",
    "n_homogeneous = hypo_hyper_meth_counts[\n",
    "    (hypo_hyper_meth_counts[\"Hypo-methylated\"] == 0)\n",
    "    | (hypo_hyper_meth_counts[\"Hyper-methylated\"] == 0)\n",
    "]\n",
    "print(\"Number of genes with only hypo or hyper methylated CpGs:\", len(n_homogeneous))\n",
    "print(\n",
    "    \"Number of genes with both hypo or hyper methylated CpGs:\",\n",
    "    len(hypo_hyper_meth_counts) - len(n_homogeneous),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46bf9b03",
   "metadata": {},
   "source": [
    "#### Metastatic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3625a1e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_meth_df = (\n",
    "    pd.read_csv(\n",
    "        (\n",
    "            wcdt_root.joinpath(\"methylkit\").joinpath(\n",
    "                \"sample_type_met_FOLH1_level_met_high+met_low__\"\n",
    "                \"diff_meth_met_high_vs_met_low_ann_genes_hg38_genes_promoters.csv\"\n",
    "            )\n",
    "        ),\n",
    "        index_col=0,\n",
    "    )\n",
    "    .replace(\"NA_character_\", np.nan)\n",
    "    .dropna(subset=[\"annot.symbol\"])\n",
    ")\n",
    "\n",
    "hypo_hyper_meth_counts = (\n",
    "    diff_meth_df.groupby(\"annot.symbol\")[\"meth.diff\"]\n",
    "    .apply(lambda x: pd.Series([(x < 0).sum(), (x >= 0).sum()]))\n",
    "    .unstack()\n",
    "    .rename(columns={0: \"Hypo-methylated\", 1: \"Hyper-methylated\"})\n",
    ")\n",
    "\n",
    "n_homogeneous = hypo_hyper_meth_counts[\n",
    "    (hypo_hyper_meth_counts[\"Hypo-methylated\"] == 0)\n",
    "    | (hypo_hyper_meth_counts[\"Hyper-methylated\"] == 0)\n",
    "]\n",
    "print(\"Number of genes with only hypo or hyper methylated CpGs:\", len(n_homogeneous))\n",
    "print(\n",
    "    \"Number of genes with both hypo or hyper methylated CpGs:\",\n",
    "    len(hypo_hyper_meth_counts) - len(n_homogeneous),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c46a94e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "diff_meth_df[\"annot.symbol\"].nunique()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.13 ('bioinfo')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "vscode": {
   "interpreter": {
    "hash": "0a3c43ae5054104da7c0ae0e54c971e03f6db1cac95975f4ca433f362d99bd9b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
